// @byted-creative/pvp-client/build/common/errorCode.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp-client/build/common/errorCode.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'errorCode', __filename);"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ERROR_CODE;
(function (ERROR_CODE) {
    ERROR_CODE[ERROR_CODE["SUCCESS"] = 0] = "SUCCESS";
    ERROR_CODE[ERROR_CODE["SERVICE_WILL_SHUT_DOWN"] = 1] = "SERVICE_WILL_SHUT_DOWN";
    ERROR_CODE[ERROR_CODE["SERVICE_NOT_RESPONSE"] = 2] = "SERVICE_NOT_RESPONSE";
    ERROR_CODE[ERROR_CODE["PARAMS_INVALID"] = 3] = "PARAMS_INVALID";
    ERROR_CODE[ERROR_CODE["MISSING_PARAM"] = 4] = "MISSING_PARAM";
    ERROR_CODE[ERROR_CODE["CONNECTION_FAIL"] = 5] = "CONNECTION_FAIL";
    ERROR_CODE[ERROR_CODE["INTERNAL_ERROR"] = 9999] = "INTERNAL_ERROR";
    ERROR_CODE[ERROR_CODE["TEAM_NOT_EXIST"] = 10000] = "TEAM_NOT_EXIST";
    ERROR_CODE[ERROR_CODE["TEAM_FULL"] = 10001] = "TEAM_FULL";
    ERROR_CODE[ERROR_CODE["PLAYER_NOT_IN_TEAM"] = 10002] = "PLAYER_NOT_IN_TEAM";
    ERROR_CODE[ERROR_CODE["PLAYER_IN_THIS_TEAM"] = 10003] = "PLAYER_IN_THIS_TEAM";
    ERROR_CODE[ERROR_CODE["PLAYER_IN_OTHER_TEAMS"] = 10004] = "PLAYER_IN_OTHER_TEAMS";
    ERROR_CODE[ERROR_CODE["NOT_MASTER"] = 10005] = "NOT_MASTER";
    ERROR_CODE[ERROR_CODE["TARGET_NOT_IN_TEAM"] = 10006] = "TARGET_NOT_IN_TEAM";
    ERROR_CODE[ERROR_CODE["CANNOT_CHANGE_CONFIG"] = 10007] = "CANNOT_CHANGE_CONFIG";
    ERROR_CODE[ERROR_CODE["CALL_CANCEL_ERROR"] = 10008] = "CALL_CANCEL_ERROR";
    ERROR_CODE[ERROR_CODE["ILLEGAL_MATCH"] = 20000] = "ILLEGAL_MATCH";
    ERROR_CODE[ERROR_CODE["ALREADY_MATCHING"] = 20001] = "ALREADY_MATCHING";
    ERROR_CODE[ERROR_CODE["MATCH_TIME_OUT"] = 20002] = "MATCH_TIME_OUT";
    ERROR_CODE[ERROR_CODE["CANNOT_CANCEL_MATCH"] = 20003] = "CANNOT_CANCEL_MATCH";
    ERROR_CODE[ERROR_CODE["ROOM_NOT_EXIST"] = 40001] = "ROOM_NOT_EXIST";
})(ERROR_CODE = exports.ERROR_CODE || (exports.ERROR_CODE = {}));
exports.remindMissingConfigWord = "please use PVPClient.setAppID or PVPClient.initConfig first";

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
