// @byted-creative/pvp-client/build/common/reqManager.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp-client/build/common/reqManager.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'reqManager', __filename);"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ReqManager = (function () {
    function ReqManager() {
        this._callbackMap = new Map();
    }
    ReqManager.prototype.reqMaker = function (cmd, data, callback) {
        var serial = this._callbackMap.size + 1;
        this._callbackMap.set(serial, callback);
        var ret = JSON.stringify({
            cmd: cmd,
            serial: serial,
            data: data,
        });
        return ret;
    };
    ReqManager.prototype.ackHandler = function (res) {
        var serial = res.serial, data = res.data, status = res.status;
        var func = this._callbackMap.get(serial);
        if (func) {
            func(status, data);
            this._callbackMap.delete(serial);
        }
    };
    return ReqManager;
}());
exports.default = ReqManager;

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
