// @byted-creative/pvp-client/build/common/userConfig.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp-client/build/common/userConfig.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'userConfig', __filename);"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var UserConfig = (function () {
    function UserConfig() {
        this.session = "";
        this.appId = "";
        this.aId = "";
        this.platformId = "";
        this.appTag = "default";
        this.extra = {};
        this.urlForPlatform = {};
    }
    return UserConfig;
}());
exports.userConfig = new UserConfig();

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
