// @byted-creative/pvp-client/build/match/startTeamMatch.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp-client/build/match/startTeamMatch.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'startTeamMatch', __filename);"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require(".");
var userConfig_1 = require("../common/userConfig");
var pvpError_1 = require("../common/pvpError");
var errorCode_1 = require("../common/errorCode");
function startTeamMatch(conf) {
    var socket = conf.socket, matchId = conf.matchId, score = conf.score;
    if (!userConfig_1.userConfig.appId) {
        throw new pvpError_1.PVPError(3, "missing appId");
    }
    var matchTask = new _1.MatchTask({
        appId: userConfig_1.userConfig.appId,
        appTag: userConfig_1.userConfig.appTag,
        session: userConfig_1.userConfig.session,
        matchId: matchId,
        score: score,
        isTeam: true,
    }, socket);
    return matchTask;
}
exports.startTeamMatch = startTeamMatch;
function genTeamMatchTask(conf) {
    var socket = conf.socket;
    if (!userConfig_1.userConfig.appId) {
        throw new pvpError_1.PVPError(errorCode_1.ERROR_CODE.MISSING_PARAM, "missing appId, " + errorCode_1.remindMissingConfigWord);
    }
    var matchTask = new _1.MatchTask({
        appId: userConfig_1.userConfig.appId,
        appTag: userConfig_1.userConfig.appTag,
        session: userConfig_1.userConfig.session || "",
        isTeam: true,
    }, socket);
    return matchTask;
}
exports.genTeamMatchTask = genTeamMatchTask;

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
