// @byted-creative/laya_pvp_ui/dist/index.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/laya_pvp_ui/dist/index.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'index', __filename);'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pvp_microphone = require('@byted-creative/pvp_microphone');
var live_ui_source = require('@byted-creative/live_ui_source');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var runtime_1 = createCommonjsModule(function (module) {
/**
 * Copyright (c) 2014-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

var runtime = (function (exports) {

  var Op = Object.prototype;
  var hasOwn = Op.hasOwnProperty;
  var undefined$1; // More compressible than void 0.
  var $Symbol = typeof Symbol === "function" ? Symbol : {};
  var iteratorSymbol = $Symbol.iterator || "@@iterator";
  var asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator";
  var toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag";

  function wrap(innerFn, outerFn, self, tryLocsList) {
    // If outerFn provided and outerFn.prototype is a Generator, then outerFn.prototype instanceof Generator.
    var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator;
    var generator = Object.create(protoGenerator.prototype);
    var context = new Context(tryLocsList || []);

    // The ._invoke method unifies the implementations of the .next,
    // .throw, and .return methods.
    generator._invoke = makeInvokeMethod(innerFn, self, context);

    return generator;
  }
  exports.wrap = wrap;

  // Try/catch helper to minimize deoptimizations. Returns a completion
  // record like context.tryEntries[i].completion. This interface could
  // have been (and was previously) designed to take a closure to be
  // invoked without arguments, but in all the cases we care about we
  // already have an existing method we want to call, so there's no need
  // to create a new function object. We can even get away with assuming
  // the method takes exactly one argument, since that happens to be true
  // in every case, so we don't have to touch the arguments object. The
  // only additional allocation required is the completion record, which
  // has a stable shape and so hopefully should be cheap to allocate.
  function tryCatch(fn, obj, arg) {
    try {
      return { type: "normal", arg: fn.call(obj, arg) };
    } catch (err) {
      return { type: "throw", arg: err };
    }
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  // Dummy constructor functions that we use as the .constructor and
  // .constructor.prototype properties for functions that return Generator
  // objects. For full spec compliance, you may wish to configure your
  // minifier not to mangle the names of these two functions.
  function Generator() {}
  function GeneratorFunction() {}
  function GeneratorFunctionPrototype() {}

  // This is a polyfill for %IteratorPrototype% for environments that
  // don't natively support it.
  var IteratorPrototype = {};
  IteratorPrototype[iteratorSymbol] = function () {
    return this;
  };

  var getProto = Object.getPrototypeOf;
  var NativeIteratorPrototype = getProto && getProto(getProto(values([])));
  if (NativeIteratorPrototype &&
      NativeIteratorPrototype !== Op &&
      hasOwn.call(NativeIteratorPrototype, iteratorSymbol)) {
    // This environment has a native %IteratorPrototype%; use it instead
    // of the polyfill.
    IteratorPrototype = NativeIteratorPrototype;
  }

  var Gp = GeneratorFunctionPrototype.prototype =
    Generator.prototype = Object.create(IteratorPrototype);
  GeneratorFunction.prototype = Gp.constructor = GeneratorFunctionPrototype;
  GeneratorFunctionPrototype.constructor = GeneratorFunction;
  GeneratorFunctionPrototype[toStringTagSymbol] =
    GeneratorFunction.displayName = "GeneratorFunction";

  // Helper for defining the .next, .throw, and .return methods of the
  // Iterator interface in terms of a single ._invoke method.
  function defineIteratorMethods(prototype) {
    ["next", "throw", "return"].forEach(function(method) {
      prototype[method] = function(arg) {
        return this._invoke(method, arg);
      };
    });
  }

  exports.isGeneratorFunction = function(genFun) {
    var ctor = typeof genFun === "function" && genFun.constructor;
    return ctor
      ? ctor === GeneratorFunction ||
        // For the native GeneratorFunction constructor, the best we can
        // do is to check its .name property.
        (ctor.displayName || ctor.name) === "GeneratorFunction"
      : false;
  };

  exports.mark = function(genFun) {
    if (Object.setPrototypeOf) {
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
    } else {
      genFun.__proto__ = GeneratorFunctionPrototype;
      if (!(toStringTagSymbol in genFun)) {
        genFun[toStringTagSymbol] = "GeneratorFunction";
      }
    }
    genFun.prototype = Object.create(Gp);
    return genFun;
  };

  // Within the body of any async function, `await x` is transformed to
  // `yield regeneratorRuntime.awrap(x)`, so that the runtime can test
  // `hasOwn.call(value, "__await")` to determine if the yielded value is
  // meant to be awaited.
  exports.awrap = function(arg) {
    return { __await: arg };
  };

  function AsyncIterator(generator, PromiseImpl) {
    function invoke(method, arg, resolve, reject) {
      var record = tryCatch(generator[method], generator, arg);
      if (record.type === "throw") {
        reject(record.arg);
      } else {
        var result = record.arg;
        var value = result.value;
        if (value &&
            typeof value === "object" &&
            hasOwn.call(value, "__await")) {
          return PromiseImpl.resolve(value.__await).then(function(value) {
            invoke("next", value, resolve, reject);
          }, function(err) {
            invoke("throw", err, resolve, reject);
          });
        }

        return PromiseImpl.resolve(value).then(function(unwrapped) {
          // When a yielded Promise is resolved, its final value becomes
          // the .value of the Promise<{value,done}> result for the
          // current iteration.
          result.value = unwrapped;
          resolve(result);
        }, function(error) {
          // If a rejected Promise was yielded, throw the rejection back
          // into the async generator function so it can be handled there.
          return invoke("throw", error, resolve, reject);
        });
      }
    }

    var previousPromise;

    function enqueue(method, arg) {
      function callInvokeWithMethodAndArg() {
        return new PromiseImpl(function(resolve, reject) {
          invoke(method, arg, resolve, reject);
        });
      }

      return previousPromise =
        // If enqueue has been called before, then we want to wait until
        // all previous Promises have been resolved before calling invoke,
        // so that results are always delivered in the correct order. If
        // enqueue has not been called before, then it is important to
        // call invoke immediately, without waiting on a callback to fire,
        // so that the async generator function has the opportunity to do
        // any necessary setup in a predictable way. This predictability
        // is why the Promise constructor synchronously invokes its
        // executor callback, and why async functions synchronously
        // execute code before the first await. Since we implement simple
        // async functions in terms of async generators, it is especially
        // important to get this right, even though it requires care.
        previousPromise ? previousPromise.then(
          callInvokeWithMethodAndArg,
          // Avoid propagating failures to Promises returned by later
          // invocations of the iterator.
          callInvokeWithMethodAndArg
        ) : callInvokeWithMethodAndArg();
    }

    // Define the unified helper method that is used to implement .next,
    // .throw, and .return (see defineIteratorMethods).
    this._invoke = enqueue;
  }

  defineIteratorMethods(AsyncIterator.prototype);
  AsyncIterator.prototype[asyncIteratorSymbol] = function () {
    return this;
  };
  exports.AsyncIterator = AsyncIterator;

  // Note that simple async functions are implemented on top of
  // AsyncIterator objects; they just return a Promise for the value of
  // the final result produced by the iterator.
  exports.async = function(innerFn, outerFn, self, tryLocsList, PromiseImpl) {
    if (PromiseImpl === void 0) PromiseImpl = Promise;

    var iter = new AsyncIterator(
      wrap(innerFn, outerFn, self, tryLocsList),
      PromiseImpl
    );

    return exports.isGeneratorFunction(outerFn)
      ? iter // If outerFn is a generator, return the full iterator.
      : iter.next().then(function(result) {
          return result.done ? result.value : iter.next();
        });
  };

  function makeInvokeMethod(innerFn, self, context) {
    var state = GenStateSuspendedStart;

    return function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        if (method === "throw") {
          throw arg;
        }

        // Be forgiving, per 25.3.3.3.3 of the spec:
        // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
        return doneResult();
      }

      context.method = method;
      context.arg = arg;

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          var delegateResult = maybeInvokeDelegate(delegate, context);
          if (delegateResult) {
            if (delegateResult === ContinueSentinel) continue;
            return delegateResult;
          }
        }

        if (context.method === "next") {
          // Setting context._sent for legacy support of Babel's
          // function.sent implementation.
          context.sent = context._sent = context.arg;

        } else if (context.method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw context.arg;
          }

          context.dispatchException(context.arg);

        } else if (context.method === "return") {
          context.abrupt("return", context.arg);
        }

        state = GenStateExecuting;

        var record = tryCatch(innerFn, self, context);
        if (record.type === "normal") {
          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          if (record.arg === ContinueSentinel) {
            continue;
          }

          return {
            value: record.arg,
            done: context.done
          };

        } else if (record.type === "throw") {
          state = GenStateCompleted;
          // Dispatch the exception by looping back around to the
          // context.dispatchException(context.arg) call above.
          context.method = "throw";
          context.arg = record.arg;
        }
      }
    };
  }

  // Call delegate.iterator[context.method](context.arg) and handle the
  // result, either by returning a { value, done } result from the
  // delegate iterator, or by modifying context.method and context.arg,
  // setting context.delegate to null, and returning the ContinueSentinel.
  function maybeInvokeDelegate(delegate, context) {
    var method = delegate.iterator[context.method];
    if (method === undefined$1) {
      // A .throw or .return when the delegate iterator has no .throw
      // method always terminates the yield* loop.
      context.delegate = null;

      if (context.method === "throw") {
        // Note: ["return"] must be used for ES3 parsing compatibility.
        if (delegate.iterator["return"]) {
          // If the delegate iterator has a return method, give it a
          // chance to clean up.
          context.method = "return";
          context.arg = undefined$1;
          maybeInvokeDelegate(delegate, context);

          if (context.method === "throw") {
            // If maybeInvokeDelegate(context) changed context.method from
            // "return" to "throw", let that override the TypeError below.
            return ContinueSentinel;
          }
        }

        context.method = "throw";
        context.arg = new TypeError(
          "The iterator does not provide a 'throw' method");
      }

      return ContinueSentinel;
    }

    var record = tryCatch(method, delegate.iterator, context.arg);

    if (record.type === "throw") {
      context.method = "throw";
      context.arg = record.arg;
      context.delegate = null;
      return ContinueSentinel;
    }

    var info = record.arg;

    if (! info) {
      context.method = "throw";
      context.arg = new TypeError("iterator result is not an object");
      context.delegate = null;
      return ContinueSentinel;
    }

    if (info.done) {
      // Assign the result of the finished delegate to the temporary
      // variable specified by delegate.resultName (see delegateYield).
      context[delegate.resultName] = info.value;

      // Resume execution at the desired location (see delegateYield).
      context.next = delegate.nextLoc;

      // If context.method was "throw" but the delegate handled the
      // exception, let the outer generator proceed normally. If
      // context.method was "next", forget context.arg since it has been
      // "consumed" by the delegate iterator. If context.method was
      // "return", allow the original .return call to continue in the
      // outer generator.
      if (context.method !== "return") {
        context.method = "next";
        context.arg = undefined$1;
      }

    } else {
      // Re-yield the result returned by the delegate method.
      return info;
    }

    // The delegate iterator is finished, so forget it and continue with
    // the outer generator.
    context.delegate = null;
    return ContinueSentinel;
  }

  // Define Generator.prototype.{next,throw,return} in terms of the
  // unified ._invoke helper method.
  defineIteratorMethods(Gp);

  Gp[toStringTagSymbol] = "Generator";

  // A Generator should always return itself as the iterator object when the
  // @@iterator function is called on it. Some browsers' implementations of the
  // iterator prototype chain incorrectly implement this, causing the Generator
  // object to not be returned from this call. This ensures that doesn't happen.
  // See https://github.com/facebook/regenerator/issues/274 for more details.
  Gp[iteratorSymbol] = function() {
    return this;
  };

  Gp.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(locs) {
    var entry = { tryLoc: locs[0] };

    if (1 in locs) {
      entry.catchLoc = locs[1];
    }

    if (2 in locs) {
      entry.finallyLoc = locs[2];
      entry.afterLoc = locs[3];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry) {
    var record = entry.completion || {};
    record.type = "normal";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryLocsList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryLocsList.forEach(pushTryEntry, this);
    this.reset(true);
  }

  exports.keys = function(object) {
    var keys = [];
    for (var key in object) {
      keys.push(key);
    }
    keys.reverse();

    // Rather than returning an object with a next method, we keep
    // things simple and return the next function itself.
    return function next() {
      while (keys.length) {
        var key = keys.pop();
        if (key in object) {
          next.value = key;
          next.done = false;
          return next;
        }
      }

      // To avoid creating an additional object, we just hang the .value
      // and .done properties off the next function object itself. This
      // also ensures that the minifier will not anonymize the function.
      next.done = true;
      return next;
    };
  };

  function values(iterable) {
    if (iterable) {
      var iteratorMethod = iterable[iteratorSymbol];
      if (iteratorMethod) {
        return iteratorMethod.call(iterable);
      }

      if (typeof iterable.next === "function") {
        return iterable;
      }

      if (!isNaN(iterable.length)) {
        var i = -1, next = function next() {
          while (++i < iterable.length) {
            if (hasOwn.call(iterable, i)) {
              next.value = iterable[i];
              next.done = false;
              return next;
            }
          }

          next.value = undefined$1;
          next.done = true;

          return next;
        };

        return next.next = next;
      }
    }

    // Return an iterator with no values.
    return { next: doneResult };
  }
  exports.values = values;

  function doneResult() {
    return { value: undefined$1, done: true };
  }

  Context.prototype = {
    constructor: Context,

    reset: function(skipTempReset) {
      this.prev = 0;
      this.next = 0;
      // Resetting context._sent for legacy support of Babel's
      // function.sent implementation.
      this.sent = this._sent = undefined$1;
      this.done = false;
      this.delegate = null;

      this.method = "next";
      this.arg = undefined$1;

      this.tryEntries.forEach(resetTryEntry);

      if (!skipTempReset) {
        for (var name in this) {
          // Not sure about the optimal order of these conditions:
          if (name.charAt(0) === "t" &&
              hasOwn.call(this, name) &&
              !isNaN(+name.slice(1))) {
            this[name] = undefined$1;
          }
        }
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;

        if (caught) {
          // If the dispatched exception was caught by a catch block,
          // then let that catch block handle the exception normally.
          context.method = "next";
          context.arg = undefined$1;
        }

        return !! caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    abrupt: function(type, arg) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") &&
            this.prev < entry.finallyLoc) {
          var finallyEntry = entry;
          break;
        }
      }

      if (finallyEntry &&
          (type === "break" ||
           type === "continue") &&
          finallyEntry.tryLoc <= arg &&
          arg <= finallyEntry.finallyLoc) {
        // Ignore the finally entry if control is not jumping to a
        // location outside the try/catch block.
        finallyEntry = null;
      }

      var record = finallyEntry ? finallyEntry.completion : {};
      record.type = type;
      record.arg = arg;

      if (finallyEntry) {
        this.method = "next";
        this.next = finallyEntry.finallyLoc;
        return ContinueSentinel;
      }

      return this.complete(record);
    },

    complete: function(record, afterLoc) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = this.arg = record.arg;
        this.method = "return";
        this.next = "end";
      } else if (record.type === "normal" && afterLoc) {
        this.next = afterLoc;
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.finallyLoc === finallyLoc) {
          this.complete(entry.completion, entry.afterLoc);
          resetTryEntry(entry);
          return ContinueSentinel;
        }
      }
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(iterable, resultName, nextLoc) {
      this.delegate = {
        iterator: values(iterable),
        resultName: resultName,
        nextLoc: nextLoc
      };

      if (this.method === "next") {
        // Deliberately forget the last sent value so that we don't
        // accidentally pass it on to the delegate.
        this.arg = undefined$1;
      }

      return ContinueSentinel;
    }
  };

  // Regardless of whether this script is executing as a CommonJS module
  // or not, return the runtime object so that we can declare the variable
  // regeneratorRuntime in the outer scope, which allows this module to be
  // injected easily by `bin/regenerator --include-runtime script.js`.
  return exports;

}(
  // If this script is executing as a CommonJS module, use module.exports
  // as the regeneratorRuntime namespace. Otherwise create a new empty
  // object. Either way, the resulting object will be used to initialize
  // the regeneratorRuntime variable at the top of this file.
   module.exports 
));

try {
  regeneratorRuntime = runtime;
} catch (accidentalStrictMode) {
  // This module should not be running in strict mode, so the above
  // assignment should always work unless something is misconfigured. Just
  // in case runtime.js accidentally runs in strict mode, we can escape
  // strict mode using a global Function call. This could conceivably fail
  // if a Content Security Policy forbids using Function, but in that case
  // the proper solution is to fix the accidental strict mode problem. If
  // you've misconfigured your bundler to force strict mode and applied a
  // CSP to forbid Function, and you're not willing to fix either of those
  // problems, please detail your unique predicament in a GitHub issue.
  Function("r", "regeneratorRuntime = r")(runtime);
}
});

class MatchUILaya {
    constructor() {
        this._avartarLayout = [
            [{ x: 0, y: 0 }],
            [{ x: 150, y: 114 }],
            [
                { x: 150, y: 135 },
                { x: 651, y: 135 },
            ],
            [
                { x: 75, y: 156 },
                { x: 420, y: 156 },
                { x: 771, y: 156 },
            ],
            [
                { x: 51, y: 156 },
                { x: 297, y: 156 },
                { x: 543, y: 156 },
                { x: 789, y: 156 },
            ],
            [
                { x: 72, y: 90 },
                { x: 420, y: 90 },
                { x: 768, y: 90 },
                { x: 243, y: 375 },
                { x: 600, y: 375 },
            ],
            [
                { x: 72, y: 90 },
                { x: 420, y: 90 },
                { x: 768, y: 90 },
                { x: 72, y: 375 },
                { x: 420, y: 375 },
                { x: 768, y: 375 },
            ],
            [
                { x: 51, y: 90 },
                { x: 297, y: 90 },
                { x: 543, y: 90 },
                { x: 795, y: 90 },
                { x: 177, y: 375 },
                { x: 426, y: 375 },
                { x: 675, y: 375 },
            ],
            [
                { x: 51, y: 90 },
                { x: 297, y: 90 },
                { x: 543, y: 90 },
                { x: 795, y: 90 },
                { x: 51, y: 375 },
                { x: 297, y: 375 },
                { x: 543, y: 375 },
                { x: 795, y: 375 },
            ],
            [
                { x: 75, y: 90 },
                { x: 420, y: 90 },
                { x: 771, y: 90 },
                { x: 75, y: 375 },
                { x: 420, y: 375 },
                { x: 771, y: 375 },
                { x: 75, y: 660 },
                { x: 420, y: 660 },
                { x: 771, y: 660 },
            ],
        ];
    }
    static get instance() {
        if (MatchUILaya._instance == null) {
            MatchUILaya._instance = new MatchUILaya();
            MatchUILaya._instance._createUI();
        }
        return MatchUILaya._instance;
    }
    _setScale(_node, _scale) {
        _node.scaleX = _scale;
        _node.scaleY = _scale;
    }
    _setLocation(_node, _x, _y) {
        _node.x = this._leftMargin + _x * this._scale;
        _node.y = this._topMargin + _y * this._scale;
    }
    _createUI() {
        console.log('MatchUI_createUI()');
        // 设计分辨率：指设计图纸分辨率
        const _designWidth = 1125;
        const _designHeight = 2001;
        this._scale =
            Laya.stage.width / _designWidth < Laya.stage.height / _designHeight
                ? Laya.stage.width / _designWidth
                : Laya.stage.height / _designHeight;
        this._leftMargin = (Laya.stage.width - _designWidth * this._scale) / 2;
        this._topMargin = (Laya.stage.height - _designHeight * this._scale) / 2;
        this._dialog = new Laya.Dialog();
        this._bgImg = new Laya.Image(live_ui_source.BACKGROUNDIMG);
        this._headbgImg = new Laya.Image(live_ui_source.AVATARBGMIN);
        this._pointImgs = new Array(5);
        this._leaveBtn = new Laya.Image(live_ui_source.LEAVEBTN);
        this._wait = new Laya.Image(live_ui_source.MATCHIMG);
        this._success = new Laya.Image(live_ui_source.MATCHSUCCESS);
        this._dialog.width = this._bgImg.width = Laya.stage.width;
        this._dialog.height = this._bgImg.height = Laya.stage.height;
        this._dialog.on(Laya.Event.CLICK, this, (e) => {
            e && e.stopPropagation && e.stopPropagation();
        });
        this._dialog.popupEffect = new Laya.Handler();
        this._dialog.closeEffect = new Laya.Handler();
        this._setScale(this._headbgImg, this._scale);
        this._setScale(this._leaveBtn, this._scale);
        this._setScale(this._wait, this._scale);
        this._setScale(this._success, this._scale);
        this._wait.anchorX = this._success.anchorX = 0.5;
        this._wait.anchorY = this._success.anchorY = 0.5;
        this._setLocation(this._headbgImg, 60, 726);
        this._setLocation(this._leaveBtn, 867, 1749);
        this._setLocation(this._wait, _designWidth / 2, 513);
        this._setLocation(this._success, _designWidth / 2, 513);
        this._closeEvent = () => {
            this.close();
        };
        this._leaveBtn.on(Laya.Event.CLICK, this, this._closeEvent);
        this._masksp = new Laya.Image(live_ui_source.WHITECIRCLE);
        for (let _i = 0; _i < 5; _i++) {
            this._pointImgs[_i] = new Laya.Image(live_ui_source.LITTLECIRCLE);
            this._pointImgs[_i].anchorX = this._pointImgs[_i].anchorY = 0.5;
            this._wait.addChild(this._pointImgs[_i]);
            this._pointImgs[_i].x = 465 + 45 * _i;
            this._pointImgs[_i].y = 615;
        }
        this._dialog.addChild(this._bgImg);
        this._dialog.addChild(this._headbgImg);
        this._dialog.addChild(this._wait);
        this._dialog.addChild(this._success);
        // this._dialog.addChild(this._leaveBtn);
        const dialogLayer = new Laya.Sprite();
        Laya.stage.addChild(dialogLayer);
        dialogLayer.addChild(Laya.Dialog.manager);
        console.log('MatchUI_createUI_End()');
    }
    _updateWaitUI() {
        const _num = this._waittime % 5;
        for (let _i = 0; _i < 5; _i++) {
            if (_i === _num) {
                this._pointImgs[_i].skin = live_ui_source.BIGCIRCLE;
            }
            else {
                this._pointImgs[_i].skin = live_ui_source.LITTLECIRCLE;
            }
        }
    }
    _updateTime() {
        this._waittime++;
        this._updateWaitUI();
    }
    // 绘制头像信息
    _updateHeads(_playerNum, _playerInfoList, _ifDrawHeads = false) {
        this._headbgImg.removeChildren();
        this._headbgImg.skin =
            _playerNum < 5
                ? live_ui_source.AVATARBGMIN
                : _playerNum < 9
                    ? live_ui_source.AVATARBGMID
                    : live_ui_source.AVATARBGMAX;
        const layout = this._avartarLayout[_playerNum];
        if (!layout) {
            return;
        }
        for (let _i = 0; _i < _playerNum; _i++) {
            const _headGroup = new Laya.Image();
            _headGroup.x = layout[_i].x;
            _headGroup.y = layout[_i].y;
            this._headbgImg.addChild(_headGroup);
            // 绘制头像
            const _headimg = new Laya.Image();
            _headimg.width = _playerNum < 3 ? 180 : 144;
            _headGroup.width = _headimg.width + (_playerNum < 3 ? 12 : 9) * 2;
            _headimg.height = _playerNum < 3 ? 180 : 144;
            this._masksp.width = _playerNum < 3 ? 180 : 144;
            this._masksp.height = _playerNum < 3 ? 180 : 144;
            _headimg.x = _playerNum < 3 ? 12 : 9;
            _headimg.y = _playerNum < 3 ? 12 : 9;
            if ((_i < _playerInfoList.length && _ifDrawHeads) || _i === 0) {
                // eslint-disable-next-line no-undef
                const localImage = new window.Image();
                localImage.crossOrigin = 'anonymous';
                localImage.src = _playerInfoList[_i].avatarUrl;
                localImage.onload = () => {
                    const scale = localImage.width / _headimg.width;
                    const t2d = new Laya.Texture2D(localImage.width / scale, localImage.height / scale, 
                    // (Laya as any).TextureFormat.R8G8B8A8
                    Laya.Texture2D.FORMAT_R8G8B8A8);
                    _headimg.graphics.drawImage(new Laya.Texture(t2d), 0, 0);
                    t2d.loadImageSource(localImage);
                    _headimg.mask = MatchUILaya.instance._masksp;
                };
            }
            else {
                _headimg.skin = live_ui_source.DEFAULTAVATAR;
            }
            _headGroup.addChild(_headimg);
            if (_i < _playerInfoList.length && _playerInfoList[_i].isAnchor) {
                const _redcircleimg = new Laya.Image(live_ui_source.REDCIRCLE);
                const _zhibozhongimg = new Laya.Image(live_ui_source.ZHIBOZHONG);
                _redcircleimg.width = _redcircleimg.height =
                    _playerNum < 3 ? 204 : 162;
                _zhibozhongimg.x = _playerNum < 3 ? 45 : 36;
                _zhibozhongimg.y = _playerNum < 3 ? 168 : 132;
                _zhibozhongimg.width = _playerNum < 3 ? 114 : 90.54;
                _zhibozhongimg.height = _playerNum < 3 ? 45 : 35.73;
                _headGroup.addChild(_redcircleimg);
                _headGroup.addChild(_zhibozhongimg);
            }
            const _namelbl = new Laya.Label();
            if (_i < _playerInfoList.length) {
                _namelbl.text = _playerInfoList[_i].nickName;
                _namelbl.alpha = 1.0;
            }
            else {
                _namelbl.text = '等待玩家';
                _namelbl.alpha = 0.5;
            }
            _namelbl.fontSize = 42;
            _namelbl.color = '#ffffff';
            _namelbl.autoSize = true;
            if (_namelbl.width > _headGroup.width * 1.2) {
                const str = _namelbl.text.toString();
                let i = 1;
                while (true) {
                    _namelbl.text = str.slice(0, i);
                    _namelbl.text = _namelbl.text + '...';
                    i++;
                    _namelbl.autoSize = true;
                    if (i === str.length ||
                        _namelbl.width > _headimg.width * 1.2) {
                        break;
                    }
                }
            }
            _headGroup.addChild(_namelbl);
            _namelbl.y = _playerNum < 3 ? 225 : 186;
            _namelbl.x = (_headGroup.width - _namelbl.width) / 2;
        }
    }
    show(_playerNum, _userInfo, _debug = false) {
        console.log('MatchUI_show()');
        // 在显示匹配页面时手动关闭连麦
        MicrophoneUILaya.instance.close();
        this._dialog.visible = true;
        Laya.timer.clear(this, this._updateTime);
        this._headbgImg.removeChildren();
        this._setScale(this._wait, this._scale);
        this._success.visible = false;
        this._wait.visible = true;
        this._waittime = 0;
        this._myOpenId = _userInfo.openId;
        this._updateHeads(_playerNum, [_userInfo], true);
        this._dialog.show();
        this._updateWaitUI();
        if (_debug) {
            this._setScale(Laya.Stat, this._scale);
            Laya.Stat.show(0, 0);
            // DebugUI.instance.show();
        }
        Laya.timer.loop(300, this, this._updateTime);
        console.log('MatchUI_show()_End');
    }
    matchSuccess(_playerList) {
        console.log('MatchUI_matchSuccess()');
        this._closePromise = new Promise(resolve => {
            this._closeResolve = resolve;
        });
        const _playerInfoList = [];
        // 把我放在第一位
        for (let i = 0; i < _playerList.length; i++) {
            if (_playerList[i].extraInfo.openId === this._myOpenId) {
                _playerInfoList.push(_playerList[i].extraInfo);
            }
        }
        // 把其他人放在后面
        for (let i = 0; i < _playerList.length; i++) {
            if (_playerList[i].extraInfo.openId !== this._myOpenId) {
                _playerInfoList.push(_playerList[i].extraInfo);
            }
        }
        this._dialog.visible = true;
        this._dialog.show();
        const _playerNum = _playerInfoList.length;
        Laya.timer.clear(this, this._updateTime);
        this._headbgImg.removeChildren();
        this._updateHeads(_playerNum, _playerInfoList, false);
        Laya.Tween.to(this._wait, { scaleX: 0.01, scaleY: 0.01 }, 500, Laya.Ease.backOut, Laya.Handler.create(this, () => {
            this._success.visible = false;
            this._success.scaleX = 0.01;
            this._success.scaleY = 0.01;
            this._success.visible = true;
            Laya.Tween.to(this._success, { scaleX: this._scale, scaleY: this._scale }, 500, Laya.Ease.backOut, Laya.Handler.create(this, () => {
                this._headbgImg.removeChildren();
                this._updateHeads(_playerNum, _playerInfoList, true);
                Laya.timer.once(3000, this, this.close);
            }));
        }));
    }
    whenClosed() {
        return this._closePromise;
    }
    setCloseEvent(_event) {
        this._closeEvent = _event;
    }
    close() {
        this._headbgImg.removeChildren();
        Laya.timer.clear(this, this._updateTime);
        this._dialog.close();
        this._dialog.visible = false;
        if (this._closeResolve) {
            this._closeResolve();
            this._closeResolve = null;
        }
    }
}
class DebugUI {
    static get instance() {
        if (DebugUI._instance == null) {
            DebugUI._instance = new DebugUI();
            DebugUI._instance._createUI();
        }
        return DebugUI._instance;
    }
    _createUI() {
        console.log('Debug_createUI()');
        this._debugbtn = new Laya.Button(undefined, 'vConsole');
        this._debugclr = new Laya.Button(undefined, 'clear');
        this._debugbtn.graphics.drawRect(0, 0, 200, 60, '#00dd00');
        this._debugclr.graphics.drawRect(0, 0, 200, 60, '#00dd00');
        this._debugbtn.labelColors = '#ffffff';
        this._debugclr.labelColors = '#ffffff';
        this._debugpnl = new Laya.Panel();
        this._debuglbl = new Laya.Text();
        this._debugbtn.size(200, 60);
        this._debugclr.size(200, 60);
        this._debugpnl.size(Laya.stage.width, Laya.stage.height);
        this._debuglbl.size(Laya.stage.width, Laya.stage.height - 60);
        Laya.stage.addChild(this._debugbtn);
        Laya.stage.addChild(this._debugpnl);
        this._debugpnl.addChild(this._debuglbl);
        this._debugpnl.addChild(this._debugclr);
        this._debugpnl.x = 0;
        this._debugpnl.y = 0;
        this._debuglbl.x = 0;
        this._debuglbl.y = 0;
        this._debugclr.x = 200;
        this._debugbtn.y = this._debugclr.y = Laya.stage.height - 60;
        this._debugbtn.zOrder = 999999999;
        this._debugpnl.zOrder = 999999998;
        this._debuglbl.bgColor = '#ffffff';
        this._debuglbl.overflow = Laya.Text.SCROLL;
        this._debuglbl.wordWrap = true;
        this._debuglbl.fontSize = 32;
        this._debugbtn.labelSize = 32;
        this._debugclr.labelSize = 32;
        this._debuglbl.on(Laya.Event.MOUSE_DOWN, this, this._startScrollText);
        this._debugbtn.on(Laya.Event.CLICK, this, this._onClickDebug);
        this._debugclr.on(Laya.Event.CLICK, this, this._onClickClear);
        this._debugbtn.visible = false;
        this._debugpnl.visible = false;
        console.log('Debug_createUI()_End');
    }
    show() {
        console.log('Debug_show()');
        this._debugbtn.visible = true;
        console.log('Debug_show()_End');
    }
    log(args) {
        console.log('tiago laya debug', ...args);
        // if (this._logString) {
        //     this._logString += '\n';
        // }
        // for (let i = 0; i < args.length; i++) {
        //     if (Array.isArray(args[i])) {
        //         const param = args[i];
        //         for (let index = 0; index < param.length; index++) {
        //             this._logString +=
        //                 typeof param[index] === 'string'
        //                     ? param[index]
        //                     : JSON.stringify(param[index], this.replace);
        //             this._seen = [];
        //             this._logString += ' ';
        //         }
        //     } else {
        //         this._logString +=
        //             typeof args[i] === 'string'
        //                 ? args[i]
        //                 : JSON.stringify(args[i], this.replace);
        //         this._seen = [];
        //         this._logString += ' ';
        //     }
        // }
        // if (this._debuglbl) {
        //     this._debuglbl.text = this._logString;
        // }
    }
    // private replace(...param: any[]) {
    //     const value = param[1];
    //     if (value != null && typeof value === 'object') {
    //         if (DebugUI.instance._seen.indexOf(value) >= 0) {
    //             return;
    //         }
    //         DebugUI.instance._seen.push(value);
    //     }
    //     return value;
    // }
    _onClickDebug() {
        this._debugpnl.visible = !this._debugpnl.visible;
        if (this._debugpnl.visible) {
            Laya.Stat.hide();
        }
        else {
            Laya.Stat.show();
        }
    }
    _onClickClear() {
        // this._logString = '';
        this._debuglbl.text = '';
    }
    /* 开始滚动文本 */
    _startScrollText() {
        this._prevX = this._debuglbl.mouseX;
        this._prevY = this._debuglbl.mouseY;
        Laya.stage.on(Laya.Event.MOUSE_MOVE, this, this._scrollText);
        Laya.stage.on(Laya.Event.MOUSE_UP, this, this._finishScrollText);
    }
    /* 停止滚动文本 */
    _finishScrollText() {
        Laya.stage.off(Laya.Event.MOUSE_MOVE, this, this._scrollText);
        Laya.stage.off(Laya.Event.MOUSE_UP, this, this._finishScrollText);
    }
    /* 鼠标滚动文本 */
    _scrollText() {
        const nowX = this._debuglbl.mouseX;
        const nowY = this._debuglbl.mouseY;
        this._debuglbl.scrollX += this._prevX - nowX;
        this._debuglbl.scrollY += this._prevY - nowY;
        this._prevX = nowX;
        this._prevY = nowY;
    }
}
class MicrophoneUILaya {
    constructor() {
        this._isMyOpen = false;
        this._isOtherOpen = false;
    }
    static get instance() {
        if (MicrophoneUILaya._instance == null) {
            MicrophoneUILaya._instance = new MicrophoneUILaya();
            MicrophoneUILaya._instance._createUI();
        }
        return MicrophoneUILaya._instance;
    }
    _updateUI() {
        if (this._isMyOpen) {
            this._mySprite.skin = live_ui_source.MICMYOPEN;
        }
        else {
            this._mySprite.skin = live_ui_source.MICMYCLOSE;
        }
        if (this._isOtherOpen) {
            this._otherSprite.skin = live_ui_source.MICOTHEROPEN;
        }
        else {
            this._otherSprite.skin = live_ui_source.MICOTHERCLOSE;
        }
    }
    _createUI() {
        console.log('MicrophoneUILaya_createUI()');
        // 设计分辨率：指设计图纸分辨率
        const _designWidth = 1125;
        const _designHeight = 2001;
        const _scale = Laya.stage.width / _designWidth < Laya.stage.height / _designHeight
            ? Laya.stage.width / _designWidth
            : Laya.stage.height / _designHeight;
        this._dialog = new Laya.Scene();
        this._mySprite = new Laya.Image();
        this._otherSprite = new Laya.Image();
        // this._dialog.isPopupCenter = false;
        this._dialog.width = 249 * _scale;
        this._dialog.height = 96 * _scale;
        this._dialog.y = 51 * _scale;
        this._dialog.x = 22; // Laya.stage.width - this._dialog.width;
        this._mySprite.width = this._mySprite.height = this._otherSprite.width = this._otherSprite.height =
            96 * _scale;
        this._otherSprite.right = 33 * _scale;
        this._updateUI();
        // this._dialog.popupEffect = new Laya.Handler();
        // this._dialog.closeEffect = new Laya.Handler();
        this._mySprite.on(Laya.Event.CLICK, this, this.clickMyChat);
        this._otherSprite.on(Laya.Event.CLICK, this, this.clickOtherChat);
        this._dialog.addChild(this._mySprite);
        this._dialog.addChild(this._otherSprite);
        this._dialog.visible = false;
        Laya.stage.addChild(this._dialog);
        this._dialog.zOrder = 999999999;
        console.log('MicrophoneUILaya_createUI()_End');
    }
    show(config) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!runtime_1) {
                return false;
            }
            console.log('MicrophoneUILaya_showUI()');
            try {
                yield pvp_microphone.PVPMicrophone.init({
                    openID: config.openID,
                    roomNum: config.roomNum,
                    appID: config.appID,
                    source: config.source,
                });
            }
            catch (error) {
                console.error('init error', error);
                this.close();
                return;
            }
            console.log('init success');
            console.log(pvp_microphone.PVPMicrophone.getDebugInfo());
            try {
                yield pvp_microphone.PVPMicrophone.startChat();
            }
            catch (e) {
                console.error('start chat error', e);
                this.close();
                return false;
            }
            console.log('startChat success');
            this._isMyOpen = pvp_microphone.PVPMicrophone.getAudioStatus().local;
            this._isOtherOpen = pvp_microphone.PVPMicrophone.getAudioStatus().remote;
            this._updateUI();
            // this._dialog.show();
            this._dialog.visible = true;
            this._dialog.active = true;
        });
    }
    close() {
        if (this._dialog) {
            console.log('close laya microphone ui');
            pvp_microphone.PVPMicrophone.endChat();
            this._dialog.visible = false;
            this._dialog.active = false;
        }
        // this._dialog.close();
    }
    clickMyChat() {
        if (this._isMyOpen) {
            pvp_microphone.PVPMicrophone.disableLocalAudio()
                .then(() => {
                this._isMyOpen = false;
                this._updateUI();
                console.log('disableLocalAudio success');
            })
                .catch((error) => {
                console.error('disableLocalAudio error', error);
            });
        }
        else {
            pvp_microphone.PVPMicrophone.enableLocalAudio()
                .then(() => {
                this._isMyOpen = true;
                this._updateUI();
                console.log('enableLocalAudio success');
            })
                .catch((error) => {
                console.error('enableLocalAudio error', error);
            });
        }
    }
    clickOtherChat() {
        if (this._isOtherOpen) {
            pvp_microphone.PVPMicrophone.muteAllRemoteAudioStream()
                .then(() => {
                this._isOtherOpen = false;
                this._updateUI();
                console.log('muteAllRemoteAudioStream success');
            })
                .catch((error) => {
                console.error('muteAllRemoteAudioStream error', error);
            });
        }
        else {
            pvp_microphone.PVPMicrophone.unmuteAllRemoteAudioStream()
                .then(() => {
                this._isOtherOpen = true;
                this._updateUI();
                console.log('unmuteAllRemoteAudioStream success');
            })
                .catch((error) => {
                console.error('unmuteAllRemoteAudioStream error', error);
            });
        }
    }
}

exports.DebugUI = DebugUI;
exports.MatchUILaya = MatchUILaya;
exports.MicrophoneUILaya = MicrophoneUILaya;

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
