// @byted-creative/tiago/dist/index.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/tiago/dist/index.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'index', __filename);'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var eventemitter3 = require('eventemitter3');
var pvp_match = require('@byted-creative/pvp_match');
var CocosMatchUI = require('@byted-creative/cocos_pvp_ui');
var CocosMatchUI__default = _interopDefault(CocosMatchUI);
var laya_pvp_ui = require('@byted-creative/laya_pvp_ui');
var stt = _interopDefault(require('@byted-creative/stt-promisify'));
var pvp_microphone = require('@byted-creative/pvp_microphone');
var pvpClient = require('@byted-creative/pvp-client');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var _a, _b;
//
var GAME_ENV;
(function (GAME_ENV) {
    GAME_ENV["Test"] = "test";
    GAME_ENV["Release"] = "release";
})(GAME_ENV || (GAME_ENV = {}));
// 匹配类型
var MATCH_TYPE;
(function (MATCH_TYPE) {
    MATCH_TYPE[MATCH_TYPE["Single"] = 1] = "Single";
    MATCH_TYPE[MATCH_TYPE["NVN"] = 2] = "NVN";
})(MATCH_TYPE || (MATCH_TYPE = {}));
// 业务场景类型
var BUSINESS_SCENE;
(function (BUSINESS_SCENE) {
    BUSINESS_SCENE["Wonderland"] = "wonderland";
    BUSINESS_SCENE["LiveRoom"] = "liveRoom";
    BUSINESS_SCENE["Unkown"] = "unknow";
})(BUSINESS_SCENE || (BUSINESS_SCENE = {}));
var Deferred = /** @class */ (function () {
    function Deferred() {
        var _this = this;
        this.promise = new Promise(function (resolve, reject) {
            Object.defineProperties(_this, {
                resolve: {
                    value: function (value) {
                        // @ts-ignore
                        _this.value = value;
                        resolve(value);
                    },
                    writable: false,
                },
                reject: {
                    value: reject,
                    writable: false,
                },
            });
        });
    }
    return Deferred;
}());
var ERROR_CODE;
(function (ERROR_CODE) {
    // 匹配错误
    ERROR_CODE[ERROR_CODE["MATCH_SUCCESS"] = 0] = "MATCH_SUCCESS";
    ERROR_CODE[ERROR_CODE["MATCH_ERROR"] = 1] = "MATCH_ERROR";
    ERROR_CODE[ERROR_CODE["MATCH_TIMEOUT"] = 2] = "MATCH_TIMEOUT";
    ERROR_CODE[ERROR_CODE["MATCH_CANCEL"] = 3] = "MATCH_CANCEL";
    ERROR_CODE[ERROR_CODE["MATCH_CLOSE"] = 4] = "MATCH_CLOSE";
    // 连麦错误
    ERROR_CODE[ERROR_CODE["GEN_CHANNEL_ERROR"] = 101] = "GEN_CHANNEL_ERROR";
    ERROR_CODE[ERROR_CODE["GET_AUTH_FAILED"] = 102] = "GET_AUTH_FAILED";
    ERROR_CODE[ERROR_CODE["JOIN_CHANNEL_FAIL"] = 103] = "JOIN_CHANNEL_FAIL";
    ERROR_CODE[ERROR_CODE["CONNECTION_LOST"] = 104] = "CONNECTION_LOST";
    ERROR_CODE[ERROR_CODE["LEAVE_CHANNEL_FAIL"] = 105] = "LEAVE_CHANNEL_FAIL";
    ERROR_CODE[ERROR_CODE["ENABLE_LOCAL_AUDIO_FAIL"] = 106] = "ENABLE_LOCAL_AUDIO_FAIL";
    ERROR_CODE[ERROR_CODE["DISABLE_LOCAL_AUDIO_FAIL"] = 107] = "DISABLE_LOCAL_AUDIO_FAIL";
    ERROR_CODE[ERROR_CODE["MUTE_ALL_REMOTE_AUDIO_FAIL"] = 108] = "MUTE_ALL_REMOTE_AUDIO_FAIL";
    ERROR_CODE[ERROR_CODE["UNMUTE_ALL_REMOTE_AUDIO_FAIL"] = 109] = "UNMUTE_ALL_REMOTE_AUDIO_FAIL";
    // 房间错误
    ERROR_CODE[ERROR_CODE["GAME_ROOM_CONNECTION_LOST"] = 201] = "GAME_ROOM_CONNECTION_LOST";
    // 内部错误
    ERROR_CODE[ERROR_CODE["WITHOUT_INIT"] = 301] = "WITHOUT_INIT";
    ERROR_CODE[ERROR_CODE["INVALID_INPUT"] = 302] = "INVALID_INPUT";
})(ERROR_CODE || (ERROR_CODE = {}));
var MATCH_ERROR_CODE;
(function (MATCH_ERROR_CODE) {
    MATCH_ERROR_CODE[MATCH_ERROR_CODE["MATCH_SUCCESS"] = 0] = "MATCH_SUCCESS";
    MATCH_ERROR_CODE[MATCH_ERROR_CODE["MATCH_ERROR"] = 1] = "MATCH_ERROR";
    MATCH_ERROR_CODE[MATCH_ERROR_CODE["MATCH_TIMEOUT"] = 2] = "MATCH_TIMEOUT";
    MATCH_ERROR_CODE[MATCH_ERROR_CODE["MATCH_CANCEL"] = 3] = "MATCH_CANCEL";
    MATCH_ERROR_CODE[MATCH_ERROR_CODE["MATCH_CLOSE"] = 4] = "MATCH_CLOSE";
})(MATCH_ERROR_CODE || (MATCH_ERROR_CODE = {}));
var MATCH_ERROR_MSG = (_a = {},
    _a[MATCH_ERROR_CODE.MATCH_SUCCESS] = 'success',
    _a[MATCH_ERROR_CODE.MATCH_ERROR] = 'match error',
    _a[MATCH_ERROR_CODE.MATCH_TIMEOUT] = 'match timeout',
    _a[MATCH_ERROR_CODE.MATCH_CANCEL] = 'match cancel',
    _a[MATCH_ERROR_CODE.MATCH_CLOSE] = 'match interrupted by other reasons',
    _a);
var RTC_ERROR_CODE;
(function (RTC_ERROR_CODE) {
    RTC_ERROR_CODE[RTC_ERROR_CODE["GEN_CHANNEL_ERROR"] = 1] = "GEN_CHANNEL_ERROR";
    RTC_ERROR_CODE[RTC_ERROR_CODE["GET_AUTH_FAILED"] = 2] = "GET_AUTH_FAILED";
    RTC_ERROR_CODE[RTC_ERROR_CODE["JOIN_CHANNEL_FAIL"] = 3] = "JOIN_CHANNEL_FAIL";
    RTC_ERROR_CODE[RTC_ERROR_CODE["CONNECTION_LOST"] = 4] = "CONNECTION_LOST";
    RTC_ERROR_CODE[RTC_ERROR_CODE["LEAVE_CHANNEL_FAIL"] = 5] = "LEAVE_CHANNEL_FAIL";
    RTC_ERROR_CODE[RTC_ERROR_CODE["ENABLE_LOCAL_AUDIO_FAIL"] = 6] = "ENABLE_LOCAL_AUDIO_FAIL";
    RTC_ERROR_CODE[RTC_ERROR_CODE["DISABLE_LOCAL_AUDIO_FAIL"] = 7] = "DISABLE_LOCAL_AUDIO_FAIL";
    RTC_ERROR_CODE[RTC_ERROR_CODE["MUTE_ALL_REMOTE_AUDIO_FAIL"] = 8] = "MUTE_ALL_REMOTE_AUDIO_FAIL";
    RTC_ERROR_CODE[RTC_ERROR_CODE["UNMUTE_ALL_REMOTE_AUDIO_FAIL"] = 9] = "UNMUTE_ALL_REMOTE_AUDIO_FAIL";
})(RTC_ERROR_CODE || (RTC_ERROR_CODE = {}));
var RTC_ERROR_MSG = (_b = {},
    _b[RTC_ERROR_CODE.GEN_CHANNEL_ERROR] = '生成连麦频道id失败',
    _b[RTC_ERROR_CODE.GET_AUTH_FAILED] = '获取权限失败',
    _b[RTC_ERROR_CODE.JOIN_CHANNEL_FAIL] = '加入连麦频道失败',
    _b[RTC_ERROR_CODE.CONNECTION_LOST] = '连麦频道链接丢失',
    _b[RTC_ERROR_CODE.LEAVE_CHANNEL_FAIL] = '离开连麦频道失败',
    _b[RTC_ERROR_CODE.ENABLE_LOCAL_AUDIO_FAIL] = '开启本地推流失败',
    _b[RTC_ERROR_CODE.DISABLE_LOCAL_AUDIO_FAIL] = '关闭本地推流失败',
    _b[RTC_ERROR_CODE.MUTE_ALL_REMOTE_AUDIO_FAIL] = '远端静音失败',
    _b[RTC_ERROR_CODE.UNMUTE_ALL_REMOTE_AUDIO_FAIL] = '解除远端静音失败',
    _b);
var GAME_ROOM_ERROR_CODE;
(function (GAME_ROOM_ERROR_CODE) {
    GAME_ROOM_ERROR_CODE[GAME_ROOM_ERROR_CODE["CONNECTION_LOST"] = 201] = "CONNECTION_LOST";
})(GAME_ROOM_ERROR_CODE || (GAME_ROOM_ERROR_CODE = {}));
var TiagoError = /** @class */ (function (_super) {
    __extends(TiagoError, _super);
    function TiagoError(code, msg) {
        var _this = _super.call(this, msg) || this;
        _this.code = code;
        _this.msg = msg;
        return _this;
    }
    return TiagoError;
}(Error));

var Match = /** @class */ (function (_super) {
    __extends(Match, _super);
    function Match(params) {
        var _this = _super.call(this) || this;
        if (params && params.debugInstance) {
            _this._debugInstance = params.debugInstance;
            pvp_match.PVPMatch.setDebugLog(_this._debugInstance);
        }
        _this._config = params;
        return _this;
    }
    Match.prototype.startMatch = function () {
        return __awaiter(this, void 0, void 0, function () {
            var initInfo, launchScene, matchUI;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initInfo = __assign({}, this._config);
                        launchScene = initInfo.launchScene || initInfo.launch_scene;
                        matchUI = this.getMatchUI();
                        initInfo.userInfo.isAnchor = initInfo.playerRole === 'streamer';
                        this.log('input match init data', {
                            gameId: initInfo.gameId,
                            environment: initInfo.environment,
                            hostAppId: initInfo.hostAppId,
                            userInfo: initInfo.userInfo,
                            isMultiPlayer: initInfo.isMultiPlayer,
                            isAutoAppendAI: initInfo.isAutoAppendAI,
                            isNewcomer: initInfo.isNewcomer,
                        });
                        pvp_match.PVPMatch.init({
                            gameId: initInfo.gameId,
                            environment: initInfo.environment,
                            hostAppId: initInfo.hostAppId,
                            userInfo: initInfo.userInfo,
                            isMultiPlayer: initInfo.isMultiPlayer,
                            isAutoAppendAI: initInfo.isAutoAppendAI,
                            isNewcomer: initInfo.isNewcomer,
                        });
                        if (!(!this._config.disableMatchUI && matchUI)) return [3 /*break*/, 2];
                        return [4 /*yield*/, matchUI.show(initInfo.playerNumber, initInfo.userInfo, this._config.debug)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        this.log('init info', initInfo);
                        if (initInfo && launchScene === 'liveStream') {
                            this.log('live sdk setPreferredFramesPerSecond 30');
                            stt.setPreferredFramesPerSecond(30);
                        }
                        pvp_match.PVPMatch.startMatch(this._config.retryCount)
                            .catch(function (error) {
                            stt.showToast({
                                title: '暂时没有对手，换个游戏玩吧',
                                icon: 'none',
                                duration: 3000
                            });
                            // todo 错误信息
                            _this.emit('error', {
                                code: error.reason,
                                msg: MATCH_ERROR_MSG[error.reason],
                            });
                        })
                            .then(function (param) { return __awaiter(_this, void 0, void 0, function () {
                            var roomNum, playerList, retTeamList, retPlayerList, i, teamInfo, extraInfo, playerInfo;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!param) {
                                            return [2 /*return*/];
                                        }
                                        this.log('match success data', param);
                                        roomNum = param.roomNum, playerList = param.playerList;
                                        retTeamList = [];
                                        retPlayerList = [];
                                        for (i = 0; i < playerList.length; i++) {
                                            teamInfo = { size: 1, playerList: [] };
                                            extraInfo = playerList[i].extraInfo;
                                            playerInfo = {
                                                openId: extraInfo.openId,
                                                nickName: extraInfo.nickName,
                                                avatarUrl: extraInfo.avatarUrl,
                                                gender: extraInfo.gender,
                                                city: extraInfo.city,
                                                province: extraInfo.province,
                                                country: extraInfo.country,
                                                language: extraInfo.language,
                                                isLiveStreaming: extraInfo.isLiveStreaming,
                                                isAI: extraInfo.isAI,
                                            };
                                            teamInfo.playerList.push(__assign({}, playerInfo));
                                            retTeamList.push(teamInfo);
                                            retPlayerList.push(__assign({}, playerInfo));
                                        }
                                        if (!(!this._config.disableMatchUI && matchUI)) return [3 /*break*/, 2];
                                        matchUI.matchSuccess(playerList);
                                        return [4 /*yield*/, matchUI.whenClosed()];
                                    case 1:
                                        _a.sent();
                                        _a.label = 2;
                                    case 2:
                                        this.emit('match-success', {
                                            id: this._config.appID + "-" + roomNum + "-" + new Date().getTime(),
                                            type: this._config.matchType,
                                            teamList: retTeamList,
                                            playerList: retPlayerList,
                                        });
                                        this.emit('create-game-room-success', { roomNum: roomNum });
                                        return [2 /*return*/];
                                }
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    Match.prototype.onMatchSuccess = function (cb) {
        this.on('match-success', cb);
    };
    Match.prototype.onCreateGameRoomSuccess = function (cb) {
        this.on('create-game-room-success', cb);
    };
    Match.prototype.onError = function (cb) {
        this.on('error', cb);
    };
    Match.prototype.log = function () {
        var param = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            param[_i] = arguments[_i];
        }
        if (this._debugInstance && this._debugInstance.log) {
            this._debugInstance.log(param);
        }
    };
    Match.prototype.getMatchUI = function () {
        if (typeof cc !== 'undefined') {
            return CocosMatchUI__default.instance;
        }
        else if (typeof Laya !== 'undefined') {
            return laya_pvp_ui.MatchUILaya.instance;
        }
        return null;
    };
    return Match;
}(eventemitter3.EventEmitter));

/**
 * 判断当前是否为字节小游戏环境
 */
function isTT() {
    return typeof tt !== "undefined" && typeof tt.getLaunchOptionsSync === "function";
}

var utils = /*#__PURE__*/Object.freeze({
    __proto__: null,
    isTT: isTT
});

/* eslint-disable @typescript-eslint/camelcase */
var AIInfos = [
    {
        nickName: "海纳百川嘿嘿",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/29c5054ec55541078423d522fc35d0cf",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "困困了",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小曹在吃饭",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱你&",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "A在路上18634",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/91be63b83cffdea848c3330ade42b533",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "笑面",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "斜杠青年是我",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "玫瑰花海",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "点点点点",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "农村来的孩子",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户23451",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "承诺3391",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "隔壁老汪",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/9aaf893a2dece2dff55b82a616c70385",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "燕子菲菲",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户271275",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "骑猪上高速888",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/2e8439eae11f804001da48609300b1d4",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "风轻云淡",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "大大大金钢",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小菜源",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "精灵之水",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/519b8421a6c3a6049614dc4e8fff1dce",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱你到永远",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户4390231",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "欢欢和乐乐",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "火山小公民",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/833424c376a478d496c3d54de1433ade",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "种桃山",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "刘自在",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "三农学校",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "长发飘飘",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/e961a64fd0355960ba3e9922ed153278",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "君子兰和一枝梅",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "无心法师",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "心诚则灵哦",
        avatarUrl: "https://weili.pstatp.com/weili/sm/426848925334503725.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "冰要融化了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/623228522403070134.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "转角错过了你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/553477068105580572.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心一百天",
        avatarUrl: "https://weili.pstatp.com/weili/sm/240089462197190945.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户934214",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587648334714896394.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户6680453",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669616943813362018.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "长春烤冷面大师",
        avatarUrl: "https://weili.pstatp.com/weili/sm/658420230293946378.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "叫姐姐",
        avatarUrl: "https://weili.pstatp.com/weili/sm/428516205869006917.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "山野",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587446892231000357.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "南城旧梦8123",
        avatarUrl: "https://weili.pstatp.com/weili/sm/384307231256215560.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "我爱洗澡",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587229592655626693.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心过好每一天779",
        avatarUrl: "https://weili.pstatp.com/weili/sm/467280991544541423.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阜阳141",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587302976466846012.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "幸福满满140",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669632689163468848.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户30702357",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587405188098556325.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "梁凉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587297066591846784.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心61405039",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587205051212497417.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阿布都-热合曼",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669634106502676910.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "孩子的天下彤彤",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587267422727569815.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "你的丹宝吖?",
        avatarUrl: "https://weili.pstatp.com/weili/sm/462528103423148101.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "?娅熹妹?",
        avatarUrl: "https://weili.pstatp.com/weili/sm/454181633347289219.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阳光28548346",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669554520758681820.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "微笑天天341",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587253842040979803.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户22463029",
        avatarUrl: "https://weili.pstatp.com/weili/sm/250591619290759212.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一切随风413",
        avatarUrl: "https://weili.pstatp.com/weili/sm/449418703594586223.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "狼行天下6666……",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587200051870564464.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "きお征服一切",
        avatarUrl: "https://weili.pstatp.com/weili/sm/342637355472650273.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "再见只是陌路人",
        avatarUrl: "https://weili.pstatp.com/weili/sm/446882688554893315.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木头人123",
        avatarUrl: "https://weili.pstatp.com/weili/sm/545342451586564123.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "雨墨墨哥哥",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587716556047646881.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "眼镜20190304",
        avatarUrl: "https://weili.pstatp.com/weili/sm/455005837571391582.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小哥哥小哥哥",
        avatarUrl: "https://weili.pstatp.com/weili/sm/256835256216453164.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "记忆中的你2838",
        avatarUrl: "https://weili.pstatp.com/weili/sm/535044185163956250.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "虞美人的晚霞",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587677952881590544.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户65011007",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587442322385797480.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱钓鱼和吃虾",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587200215079321785.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "鹏鹏",
        avatarUrl: "https://weili.pstatp.com/weili/sm/662748749878132739.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一只小飞侠",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587258240087490915.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "珍爱一生",
        avatarUrl: "https://weili.pstatp.com/weili/sm/601065365875196168.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "何必要这样",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669651629969244198.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "我是你的小迷妹",
        avatarUrl: "https://weili.pstatp.com/weili/sm/467259224650285091.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "老男孩12345",
        avatarUrl: "https://weili.pstatp.com/weili/sm/652397371889549317.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "想家的天天",
        avatarUrl: "https://weili.pstatp.com/weili/sm/646430476903514114.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "北漂23456",
        avatarUrl: "https://weili.pstatp.com/weili/sm/503537499081670669.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "影子91",
        avatarUrl: "https://weili.pstatp.com/weili/sm/529663570392121354.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "春梦随我飞",
        avatarUrl: "https://weili.pstatp.com/weili/sm/237053212376694855.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "AAA",
        avatarUrl: "https://weili.pstatp.com/weili/sm/431201917579689985.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "DAR",
        avatarUrl: "https://weili.pstatp.com/weili/sm/538631470695186531.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户43670985",
        avatarUrl: "https://weili.pstatp.com/weili/sm/583932036957733026.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户16888150",
        avatarUrl: "https://weili.pstatp.com/weili/sm/247417140260634798.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "福282",
        avatarUrl: "https://weili.pstatp.com/weili/sm/563093439754928337.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "发哥天天发",
        avatarUrl: "https://weili.pstatp.com/weili/sm/519550932457881629.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "又来了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/447455783281230332.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "谁懂姐的心",
        avatarUrl: "https://weili.pstatp.com/weili/sm/469145531281440785.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木子81060",
        avatarUrl: "https://weili.pstatp.com/weili/sm/624045725830480155.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "user946329",
        avatarUrl: "https://weili.pstatp.com/weili/sm/554989850074284301.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "彼岸微凉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/535046461497671779.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "A偶然与巧合",
        avatarUrl: "https://weili.pstatp.com/weili/sm/555735224466931864.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "无情波斯喵",
        avatarUrl: "https://weili.pstatp.com/weili/sm/415721266307530761.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木美儿",
        avatarUrl: "https://weili.pstatp.com/weili/sm/508763202967699481.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心开心开心极了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/495360233532686349.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "★触不可及",
        avatarUrl: "https://weili.pstatp.com/weili/sm/268082624513441827.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "苗文玉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/589110118248349828.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱爱你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/628507294907891883.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "萍水相逢1234",
        avatarUrl: "https://weili.pstatp.com/weili/sm/664556407126097929.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "改变生活",
        avatarUrl: "https://weili.pstatp.com/weili/sm/580935171982229708.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "Www33",
        avatarUrl: "https://weili.pstatp.com/weili/sm/646969907616415745.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "青涩的雨",
        avatarUrl: "https://weili.pstatp.com/weili/sm/381388989957275648.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一笑而过是你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/658167626086875143.jpg?app_id=wi9fe0c5a58ce0fecd"
    }
];

var Tiago = /** @class */ (function () {
    function Tiago() {
    }
    Tiago.init = function (param) {
        return __awaiter(this, void 0, void 0, function () {
            var defer, debugInstance, currentScene, isNewcomer, queryAIInfo, userInfo;
            return __generator(this, function (_a) {
                defer = new Deferred();
                if (!this._instance) {
                    this._instance = new Tiago();
                }
                debugInstance = this.getDebugUI();
                if (param.debug && debugInstance) {
                    this._debugInstance = debugInstance;
                    pvp_microphone.PVPMicrophone.setDebugLog(debugInstance);
                }
                // eslint-disable-next-line no-undef
                if (this.utils.isTT()) {
                    this._query = tt.getLaunchOptionsSync().query;
                    console.log("获取query", this._query);
                }
                // 非tt环境，或不在W大厅和直播中，mock query
                if (!this.utils.isTT() || !this._query.launchScene) {
                    this._query = {
                        "userInfo": {
                            "avatarUrl": "http://p3.pstatp.com/thumb/379/5336884463",
                            "nickName": "测试用户",
                            "gender": 1,
                            "openId": "mock_openId_" + Math.floor(Math.random() * 100000000),
                        },
                        "playerRole": "unknown",
                        "hostAppId": "",
                        "gameId": "",
                        "playerNumber": 2,
                        "launchScene": "unknown",
                        "environment": "test",
                        // 新手局相关字段，与原 W 小游戏设计相同
                        "newcomer": false,
                    };
                    console.log("mock query", this._query);
                }
                console.log("query", this._query);
                currentScene = BUSINESS_SCENE.Unkown;
                isNewcomer = this._query.newcomer;
                queryAIInfo = this._query.aiInfo || {};
                if (this._query.launchScene === 'liveStream') {
                    currentScene = BUSINESS_SCENE.LiveRoom;
                    this._config = {
                        currentScene: currentScene,
                        scene: {
                            liveRoom: {
                                isNewcomer: isNewcomer,
                            },
                        },
                    };
                    this._aiInfo = isNewcomer ? {
                        openId: queryAIInfo.openId || queryAIInfo.open_id,
                        nickName: queryAIInfo.nickName,
                        avatarUrl: queryAIInfo.avatarUrl,
                        gender: queryAIInfo.gender,
                        city: '',
                        province: '',
                        country: '',
                        language: '',
                        isLiveStreaming: true,
                        isAI: false,
                    } : undefined;
                }
                else if (this._query.launchScene === 'wonderland') {
                    currentScene = BUSINESS_SCENE.Wonderland;
                    this._config = {
                        currentScene: currentScene,
                        scene: {
                            wonderland: {
                                isNewcomer: this._query.newcomer
                            },
                        },
                    };
                    this._aiInfo = isNewcomer ? {
                        openId: queryAIInfo.openId || queryAIInfo.open_id,
                        nickName: queryAIInfo.nickName,
                        avatarUrl: queryAIInfo.avatarUrl,
                        gender: queryAIInfo.gender,
                        city: '',
                        province: '',
                        country: '',
                        language: '',
                        isLiveStreaming: false,
                        isAI: false,
                    } : undefined;
                }
                else {
                    // todo 其他场景情况，获取信息和翻译
                    currentScene = BUSINESS_SCENE.Unkown;
                    this._config = {
                        currentScene: currentScene,
                        scene: {
                            unknow: {}
                        },
                    };
                    this._aiInfo = undefined;
                }
                this._innerConfigForV1 = {
                    environment: this._query.environment,
                    hostAppId: this._query.hostAppId,
                    launchScene: this._query.launchScene === 'liveRoom' ? 'liveStream' : 'wonderland',
                };
                userInfo = this._query.userInfo || {};
                this._userInfo = {
                    openId: userInfo.openId || userInfo.open_id,
                    nickName: userInfo.nickName,
                    avatarUrl: userInfo.avatarUrl,
                    gender: userInfo.gender,
                    city: '',
                    province: '',
                    country: '',
                    language: '',
                    isLiveStreaming: currentScene === 'liveRoom',
                };
                this._options = param;
                this._hasInit = true;
                defer.resolve();
                return [2 /*return*/, defer.promise];
            });
        });
    };
    Tiago.getMatchUI = function () {
        if (typeof cc !== 'undefined') {
            return CocosMatchUI__default.instance;
        }
        else if (typeof Laya !== 'undefined') {
            return laya_pvp_ui.MatchUILaya.instance;
        }
        return null;
    };
    Tiago.getConfig = function () {
        if (!this._hasInit) {
            throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
        }
        return this._config;
    };
    Tiago.getUserInfo = function () {
        if (!this._hasInit) {
            throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
        }
        return this._userInfo;
    };
    Tiago.closeMatchUI = function () {
        var matchUI = this.getMatchUI();
        if (matchUI) {
            matchUI.close();
        }
    };
    Tiago.startSingleMatch = function (params) {
        if (params === void 0) { params = { retryCount: 3, isAutoAppendAI: false }; }
        if (!this._hasInit) {
            throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
        }
        // 关闭麦克风
        var microUI = this.getMicrophoneUI();
        if (microUI) {
            microUI.close();
        }
        var match = new Match({
            appID: this._options.appId,
            environment: this._innerConfigForV1.environment === 'online' ? 'online' : 'test',
            playerRole: this._userInfo.isLiveStreaming ? 'streamer' : 'unknown',
            hostAppId: this._innerConfigForV1.hostAppId,
            gameId: this._options.appId,
            playerNumber: 2,
            launchScene: this._innerConfigForV1.launchScene,
            isMultiPlayer: false,
            userInfo: this._userInfo,
            debugInstance: this._debugInstance,
            debug: this._options.debug || false,
            retryCount: params.retryCount,
            matchType: this.MATCH_TYPE.NVN,
            isAutoAppendAI: params.isAutoAppendAI,
        });
        match.startMatch();
        return match;
    };
    Tiago.createNewcomerGameRoom = function () {
        if (!this._hasInit) {
            throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
        }
        var defer = new Deferred();
        // 关闭麦克风
        var microUI = this.getMicrophoneUI();
        if (microUI) {
            microUI.close();
        }
        var match = new Match({
            appID: this._options.appId,
            environment: this._innerConfigForV1.environment === 'online' ? 'online' : 'test',
            playerRole: this._userInfo.isLiveStreaming ? 'streamer' : 'unknown',
            hostAppId: this._innerConfigForV1.hostAppId,
            gameId: this._options.appId,
            playerNumber: 2,
            launchScene: this._innerConfigForV1.launchScene,
            isMultiPlayer: false,
            userInfo: this._userInfo,
            debugInstance: this._debugInstance,
            debug: this._options.debug || false,
            retryCount: 3,
            matchType: this.MATCH_TYPE.NVN,
            isNewcomer: true,
            disableMatchUI: true,
        });
        match.startMatch();
        var subDefer1 = new Deferred();
        var subDefer2 = new Deferred();
        match.on('match-success', function (param) {
            subDefer1.resolve(param);
        });
        match.on('create-game-room-success', function (param) {
            subDefer2.resolve(param);
        });
        Promise.all([subDefer1.promise, subDefer2.promise]).then(function (values) {
            defer.resolve(values[1].roomNum);
        }).catch(function (err) {
            defer.reject(err);
        });
        return defer.promise;
    };
    Tiago.joinGameRoom = function (param) {
        if (!this._hasInit) {
            throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
        }
        this._roomNum = param.roomNum;
        pvpClient.PVPClient.initConfig({
            appId: this._options.appId,
            environment: this._innerConfigForV1.environment === 'online' ? 'online' : 'test',
            aId: this._innerConfigForV1.hostAppId,
            platformId: 'wonderland',
            extraInfo: __assign(__assign({}, this._userInfo), param.extraInfo),
        });
        var room = pvpClient.PVPClient.joinGameRoom({ roomNum: param.roomNum });
        return room;
    };
    /**
     * 开始房间连麦
     */
    Tiago.joinRTCForGameRoom = function (room) {
        return __awaiter(this, void 0, void 0, function () {
            var defer, microUI;
            return __generator(this, function (_a) {
                if (!this._hasInit) {
                    throw new TiagoError(ERROR_CODE.WITHOUT_INIT, "please invoke init() before!");
                }
                defer = new Deferred();
                if (!room) {
                    defer.reject(new TiagoError(ERROR_CODE.INVALID_INPUT, "invalid input"));
                }
                microUI = this.getMicrophoneUI();
                if (microUI) {
                    this.log('input open micro data', {
                        openID: this._userInfo.openId,
                        roomNum: this._roomNum,
                        appID: this._options.appId,
                        source: 'live',
                    });
                    try {
                        microUI.show({
                            openID: this._userInfo.openId,
                            roomNum: this._roomNum,
                            appID: this._options.appId,
                            source: 'live',
                        });
                    }
                    catch (error) {
                        this.log('open micro error', error);
                    }
                }
                defer.resolve();
                return [2 /*return*/, defer.promise];
            });
        });
    };
    Tiago.getMicrophoneUI = function () {
        if (typeof cc !== 'undefined') {
            return CocosMatchUI.MicrophoneUI;
        }
        else if (typeof Laya !== 'undefined') {
            return laya_pvp_ui.MicrophoneUILaya.instance;
        }
        return null;
    };
    Tiago.getDebugUI = function () {
        if (typeof cc !== 'undefined') {
            return CocosMatchUI.DebugUI;
        }
        else if (typeof Laya !== 'undefined') {
            return laya_pvp_ui.DebugUI.instance;
        }
        return null;
    };
    Tiago.log = function () {
        var param = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            param[_i] = arguments[_i];
        }
        console.log(param);
        if (this._debugInstance) {
            this._debugInstance.log(param);
        }
    };
    Tiago.getInnerAIInfo = function () {
        var innerAIInfo = AIInfos.slice()[Math.floor(Math.random() * AIInfos.slice().length)];
        return {
            openId: innerAIInfo.open_id || '',
            nickName: innerAIInfo.nickName,
            avatarUrl: innerAIInfo.avatarUrl,
            gender: innerAIInfo.gender,
            city: innerAIInfo.city,
            province: innerAIInfo.province,
            country: innerAIInfo.country,
            language: innerAIInfo.language,
            isLiveStreaming: false,
            isAI: true,
        };
    };
    Tiago.GAME_ENV = GAME_ENV;
    Tiago.MATCH_TYPE = MATCH_TYPE;
    Tiago.BUSINESS_SCENE = BUSINESS_SCENE;
    Tiago.utils = __assign(__assign({}, utils), { getTrainerUserInfo: function () {
            return Tiago._aiInfo || Tiago.getInnerAIInfo();
        } });
    Tiago.ERROR_CODE = MATCH_ERROR_CODE;
    // public static mock: {
    //     mockConfig: (param: {
    //         currentScene: BUSINESS_SCENE,
    //         sceneParams: any,
    //     }) => void,
    //     PARAMS_TEMPLATE: typeof PARAMS_TEMPLATE;
    // };
    Tiago._hasInit = false;
    return Tiago;
}());

exports.default = Tiago;

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
