// @byted-creative/pvp_match/dist/index.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp_match/dist/index.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'index', __filename);(function (global){
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pvpClient = require('@byted-creative/pvp-client');
var errorCode = require('@byted-creative/pvp-client/build/common/errorCode');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

/**
 * 延迟承诺类。属于Promise的衍生类。
 */
var Deferred = /** @class */ (function () {
    function Deferred() {
        var _this = this;
        this.promise = new Promise(function (resolve, reject) {
            Object.defineProperties(_this, {
                resolve: {
                    value: function (value) {
                        // @ts-ignore
                        _this.value = value;
                        resolve(value);
                    },
                    writable: false,
                },
                reject: {
                    value: reject,
                    writable: false,
                },
            });
        });
    }
    return Deferred;
}());

var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var lib = createCommonjsModule(function (module, exports) {
!function(t,e){e(exports);}(commonjsGlobal,function(t){function I(t){return (I="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t})(t)}function o(t,e){if(!(t instanceof e))throw new TypeError("Cannot call a class as a function")}function r(t,e){for(var n=0;n<e.length;n++){var r=e[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(t,r.key,r);}}function i(t,e,n){return e&&r(t.prototype,e),n&&r(t,n),t}function a(){return (a=Object.assign||function(t){for(var e=1;e<arguments.length;e++){var n=arguments[e];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(t[r]=n[r]);}return t}).apply(this,arguments)}function u(t,e){if("function"!=typeof e&&null!==e)throw new TypeError("Super expression must either be null or a function");t.prototype=Object.create(e&&e.prototype,{constructor:{value:t,writable:!0,configurable:!0}}),e&&s(t,e);}function c(t){return (c=Object.setPrototypeOf?Object.getPrototypeOf:function(t){return t.__proto__||Object.getPrototypeOf(t)})(t)}function s(t,e){return (s=Object.setPrototypeOf||function(t,e){return t.__proto__=e,t})(t,e)}function l(t,e,n){return (l=function(){if("undefined"==typeof Reflect||!Reflect.construct)return !1;if(Reflect.construct.sham)return !1;if("function"==typeof Proxy)return !0;try{return Date.prototype.toString.call(Reflect.construct(Date,[],function(){})),!0}catch(t){return !1}}()?Reflect.construct:function(t,e,n){var r=[null];r.push.apply(r,e);var o=new(Function.bind.apply(t,r));return n&&s(o,n.prototype),o}).apply(null,arguments)}function e(t){var r="function"==typeof Map?new Map:void 0;return (e=function(t){if(null===t||(e=t,-1===Function.toString.call(e).indexOf("[native code]")))return t;var e;if("function"!=typeof t)throw new TypeError("Super expression must either be null or a function");if(void 0!==r){if(r.has(t))return r.get(t);r.set(t,n);}function n(){return l(t,arguments,c(this).constructor)}return n.prototype=Object.create(t.prototype,{constructor:{value:n,enumerable:!1,writable:!0,configurable:!0}}),s(n,t)})(t)}function f(t){if(void 0===t)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return t}function p(t,e){return !e||"object"!=typeof e&&"function"!=typeof e?f(t):e}var n=function(t,e){return (n=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(t,e){t.__proto__=e;}||function(t,e){for(var n in e)e.hasOwnProperty(n)&&(t[n]=e[n]);})(t,e)};var h=function(){return (h=Object.assign||function(t){for(var e,n=1,r=arguments.length;n<r;n++)for(var o in e=arguments[n])Object.prototype.hasOwnProperty.call(e,o)&&(t[o]=e[o]);return t}).apply(this,arguments)};function v(t,e){var n={};for(var r in t)Object.prototype.hasOwnProperty.call(t,r)&&e.indexOf(r)<0&&(n[r]=t[r]);if(null!=t&&"function"==typeof Object.getOwnPropertySymbols){var o=0;for(r=Object.getOwnPropertySymbols(t);o<r.length;o++)e.indexOf(r[o])<0&&Object.prototype.propertyIsEnumerable.call(t,r[o])&&(n[r[o]]=t[r[o]]);}return n}function _(i,a,s,u){return new(s=s||Promise)(function(t,e){function n(t){try{o(u.next(t));}catch(t){e(t);}}function r(t){try{o(u.throw(t));}catch(t){e(t);}}function o(e){e.done?t(e.value):new s(function(t){t(e.value);}).then(n,r);}o((u=u.apply(i,a||[])).next());})}function d(n,r){var o,i,a,t,s={label:0,sent:function(){if(1&a[0])throw a[1];return a[1]},trys:[],ops:[]};return t={next:e(0),throw:e(1),return:e(2)},"function"==typeof Symbol&&(t[Symbol.iterator]=function(){return this}),t;function e(e){return function(t){return function(e){if(o)throw new TypeError("Generator is already executing.");for(;s;)try{if(o=1,i&&(a=2&e[0]?i.return:e[0]?i.throw||((a=i.return)&&a.call(i),0):i.next)&&!(a=a.call(i,e[1])).done)return a;switch(i=0,a&&(e=[2&e[0],a.value]),e[0]){case 0:case 1:a=e;break;case 4:return s.label++,{value:e[1],done:!1};case 5:s.label++,i=e[1],e=[0];continue;case 7:e=s.ops.pop(),s.trys.pop();continue;default:if(!(a=0<(a=s.trys).length&&a[a.length-1])&&(6===e[0]||2===e[0])){s=0;continue}if(3===e[0]&&(!a||e[1]>a[0]&&e[1]<a[3])){s.label=e[1];break}if(6===e[0]&&s.label<a[1]){s.label=a[1],a=e;break}if(a&&s.label<a[2]){s.label=a[2],s.ops.push(e);break}a[2]&&s.ops.pop(),s.trys.pop();continue}e=r.call(n,s);}catch(t){e=[6,t],i=0;}finally{o=a=0;}if(5&e[0])throw e[1];return {value:e[0]?e[1]:void 0,done:!0}}([e,t])}}}function m(t){return "object"==typeof t&&null!==t&&!b(t)}function y(t){return "function"==typeof t}function g(t){return "[object String]"===Object.prototype.toString.call(t)}function b(t){return "[object Array]"===Object.prototype.toString.call(t)}function w(t){return "number"==typeof t}var x=(C.prototype._send=function(t){},C.prototype.setupTransport=function(t){var e=this;this._transport=t,this._sendQueue.forEach(function(t){e._sendEvent(t);}),this._sendQueue=[];},C);function C(t){var n=this;this._setupMonitors=function(t){return t.forEach(function(t){t.setup(n._sendEvent);}),t},this._sendEvent=function(t){if(n._transport){var e=n._modifyEvent(t);n._shouldSend(e)&&n._idleSendEvent(e);}else n._sendQueue.push(t);},this._modifyEvent=function(t){return t},this._shouldSend=function(t){return !0},this._getIdle=function(){return y(window.requestIdleCallback)?window.requestIdleCallback:y(setTimeout)?setTimeout:function(t){t();}},this._idleSendEvent=function(t){n._getIdle()(function(){n._send(t);});},this._options=t,this._sendQueue=[],this._monitors=this._setupMonitors(t.defaultMonitors||[]);}function P(t,e){var n,r,o,i;if(void 0===t.length)for(n in t)o=t,i=n,Object.prototype.hasOwnProperty.call(o,i)&&e.call(null,n,t[n]);else if(r=t.length)for(n=0;n<r;n++)e.call(null,n,t[n]);}function O(){for(var t=[],e=0;e<arguments.length;e++)t[e]=arguments[e];for(var n={},r=0;r<t.length;)n=E(n,t[r]),r++;return n}function E(t,e){for(var n in e)Object.prototype.hasOwnProperty.call(e,n)&&void 0!==e[n]&&(m(t[n])&&m(e[n])?E(t[n],e[n]):!m(t[n])&&m(e[n])?t[n]=E({},e[n]):t[n]=e[n]);return t}function S(t){var n,e=function(t){if(!m(t))return {};var n={};return P(t,function(t,e){m(e)||b(e)?n[t]=JSON.stringify(e):n[t]=e;}),n}(t),r=(n={},P(e,function(t,e){n[encodeURIComponent(t)]=encodeURIComponent(e);}),n),o=[];return P(r,function(t,e){o.push(t+"="+e);}),o.join("&")}function j(){var t=function(){for(var t=new Array(16),e=0,n=void 0;e<16;e++)0==(3&e)&&(n=4294967296*Math.random()),t[e]=n>>>((3&e)<<3)&255;return t}();return t[6]=15&t[6]|64,t[8]=63&t[8]|128,function(t){for(var e=[],n=0;n<256;++n)e[n]=(n+256).toString(16).substr(1);var r=0,o=e;return [o[t[r++]],o[t[r++]],o[t[r++]],o[t[r++]],"-",o[t[r++]],o[t[r++]],"-",o[t[r++]],o[t[r++]],"-",o[t[r++]],o[t[r++]],"-",o[t[r++]],o[t[r++]],o[t[r++]],o[t[r++]],o[t[r++]],o[t[r++]]].join("")}(t)}var k,T,L,q,M="i.snssdk.com",D=j(),A=j();function F(){return "undefined"!=typeof tt}F()&&(k=tt.request,T=tt.downloadFile,tt.uploadFile,L=tt.onError,q=tt.login);var R=k,z=T,G=L,N=q;function Q(t){return t&&t.__esModule&&Object.prototype.hasOwnProperty.call(t,"default")?t.default:t}function U(t,e){return t(e={exports:{}},e.exports),e.exports}function B(i,a,s){return new Promise(function(n,r){if("number"!=typeof a||a<0)throw new TypeError("Expected `milliseconds` to be a positive number");if(a!==1/0){var t,e,o=setTimeout(function(){if("function"!=typeof s){var t="string"==typeof s?s:"Promise timed out after ".concat(a," milliseconds"),e=s instanceof Error?s:new W(t);"function"==typeof i.cancel&&i.cancel(),r(e);}else try{n(s());}catch(t){r(t);}},a);t=i.then(n,r),e=(e=function(){clearTimeout(o);})||function(){},t.then(function(t){return new Promise(function(t){t(e());}).then(function(){return t})},function(t){return new Promise(function(t){t(e());}).then(function(){throw t})});}else n(i);})}var H=U(function(t){var e=function(i){var u,t=Object.prototype,c=t.hasOwnProperty,e="function"==typeof Symbol?Symbol:{},o=e.iterator||"@@iterator",n=e.asyncIterator||"@@asyncIterator",r=e.toStringTag||"@@toStringTag";function a(t,e,n,r){var i,a,s,u,o=e&&e.prototype instanceof m?e:m,c=Object.create(o.prototype),l=new S(r||[]);return c._invoke=(i=t,a=n,s=l,u=p,function(t,e){if(u===v)throw new Error("Generator is already running");if(u===_){if("throw"===t)throw e;return k()}for(s.method=t,s.arg=e;;){var n=s.delegate;if(n){var r=P(n,s);if(r){if(r===d)continue;return r}}if("next"===s.method)s.sent=s._sent=s.arg;else if("throw"===s.method){if(u===p)throw u=_,s.arg;s.dispatchException(s.arg);}else "return"===s.method&&s.abrupt("return",s.arg);u=v;var o=f(i,a,s);if("normal"===o.type){if(u=s.done?_:h,o.arg===d)continue;return {value:o.arg,done:s.done}}"throw"===o.type&&(u=_,s.method="throw",s.arg=o.arg);}}),c}function f(t,e,n){try{return {type:"normal",arg:t.call(e,n)}}catch(t){return {type:"throw",arg:t}}}i.wrap=a;var p="suspendedStart",h="suspendedYield",v="executing",_="completed",d={};function m(){}function s(){}function l(){}var y={};y[o]=function(){return this};var g=Object.getPrototypeOf,b=g&&g(g(j([])));b&&b!==t&&c.call(b,o)&&(y=b);var w=l.prototype=m.prototype=Object.create(y);function x(t){["next","throw","return"].forEach(function(e){t[e]=function(t){return this._invoke(e,t)};});}function C(u){var e;this._invoke=function(n,r){function t(){return new Promise(function(t,e){!function e(t,n,r,o){var i=f(u[t],u,n);if("throw"!==i.type){var a=i.arg,s=a.value;return s&&"object"===I(s)&&c.call(s,"__await")?Promise.resolve(s.__await).then(function(t){e("next",t,r,o);},function(t){e("throw",t,r,o);}):Promise.resolve(s).then(function(t){a.value=t,r(a);},function(t){return e("throw",t,r,o)})}o(i.arg);}(n,r,t,e);})}return e=e?e.then(t,t):t()};}function P(t,e){var n=t.iterator[e.method];if(n===u){if(e.delegate=null,"throw"===e.method){if(t.iterator.return&&(e.method="return",e.arg=u,P(t,e),"throw"===e.method))return d;e.method="throw",e.arg=new TypeError("The iterator does not provide a 'throw' method");}return d}var r=f(n,t.iterator,e.arg);if("throw"===r.type)return e.method="throw",e.arg=r.arg,e.delegate=null,d;var o=r.arg;return o?o.done?(e[t.resultName]=o.value,e.next=t.nextLoc,"return"!==e.method&&(e.method="next",e.arg=u),e.delegate=null,d):o:(e.method="throw",e.arg=new TypeError("iterator result is not an object"),e.delegate=null,d)}function O(t){var e={tryLoc:t[0]};1 in t&&(e.catchLoc=t[1]),2 in t&&(e.finallyLoc=t[2],e.afterLoc=t[3]),this.tryEntries.push(e);}function E(t){var e=t.completion||{};e.type="normal",delete e.arg,t.completion=e;}function S(t){this.tryEntries=[{tryLoc:"root"}],t.forEach(O,this),this.reset(!0);}function j(e){if(e){var t=e[o];if(t)return t.call(e);if("function"==typeof e.next)return e;if(!isNaN(e.length)){var n=-1,r=function t(){for(;++n<e.length;)if(c.call(e,n))return t.value=e[n],t.done=!1,t;return t.value=u,t.done=!0,t};return r.next=r}}return {next:k}}function k(){return {value:u,done:!0}}return s.prototype=w.constructor=l,l.constructor=s,l[r]=s.displayName="GeneratorFunction",i.isGeneratorFunction=function(t){var e="function"==typeof t&&t.constructor;return !!e&&(e===s||"GeneratorFunction"===(e.displayName||e.name))},i.mark=function(t){return Object.setPrototypeOf?Object.setPrototypeOf(t,l):(t.__proto__=l,r in t||(t[r]="GeneratorFunction")),t.prototype=Object.create(w),t},i.awrap=function(t){return {__await:t}},x(C.prototype),C.prototype[n]=function(){return this},i.AsyncIterator=C,i.async=function(t,e,n,r){var o=new C(a(t,e,n,r));return i.isGeneratorFunction(e)?o:o.next().then(function(t){return t.done?t.value:o.next()})},x(w),w[r]="Generator",w[o]=function(){return this},w.toString=function(){return "[object Generator]"},i.keys=function(n){var r=[];for(var t in n)r.push(t);return r.reverse(),function t(){for(;r.length;){var e=r.pop();if(e in n)return t.value=e,t.done=!1,t}return t.done=!0,t}},i.values=j,S.prototype={constructor:S,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=u,this.done=!1,this.delegate=null,this.method="next",this.arg=u,this.tryEntries.forEach(E),!t)for(var e in this)"t"===e.charAt(0)&&c.call(this,e)&&!isNaN(+e.slice(1))&&(this[e]=u);},stop:function(){this.done=!0;var t=this.tryEntries[0].completion;if("throw"===t.type)throw t.arg;return this.rval},dispatchException:function(n){if(this.done)throw n;var r=this;function t(t,e){return i.type="throw",i.arg=n,r.next=t,e&&(r.method="next",r.arg=u),!!e}for(var e=this.tryEntries.length-1;0<=e;--e){var o=this.tryEntries[e],i=o.completion;if("root"===o.tryLoc)return t("end");if(o.tryLoc<=this.prev){var a=c.call(o,"catchLoc"),s=c.call(o,"finallyLoc");if(a&&s){if(this.prev<o.catchLoc)return t(o.catchLoc,!0);if(this.prev<o.finallyLoc)return t(o.finallyLoc)}else if(a){if(this.prev<o.catchLoc)return t(o.catchLoc,!0)}else {if(!s)throw new Error("try statement without catch or finally");if(this.prev<o.finallyLoc)return t(o.finallyLoc)}}}},abrupt:function(t,e){for(var n=this.tryEntries.length-1;0<=n;--n){var r=this.tryEntries[n];if(r.tryLoc<=this.prev&&c.call(r,"finallyLoc")&&this.prev<r.finallyLoc){var o=r;break}}o&&("break"===t||"continue"===t)&&o.tryLoc<=e&&e<=o.finallyLoc&&(o=null);var i=o?o.completion:{};return i.type=t,i.arg=e,o?(this.method="next",this.next=o.finallyLoc,d):this.complete(i)},complete:function(t,e){if("throw"===t.type)throw t.arg;return "break"===t.type||"continue"===t.type?this.next=t.arg:"return"===t.type?(this.rval=this.arg=t.arg,this.method="return",this.next="end"):"normal"===t.type&&e&&(this.next=e),d},finish:function(t){for(var e=this.tryEntries.length-1;0<=e;--e){var n=this.tryEntries[e];if(n.finallyLoc===t)return this.complete(n.completion,n.afterLoc),E(n),d}},catch:function(t){for(var e=this.tryEntries.length-1;0<=e;--e){var n=this.tryEntries[e];if(n.tryLoc===t){var r=n.completion;if("throw"===r.type){var o=r.arg;E(n);}return o}}throw new Error("illegal catch attempt")},delegateYield:function(t,e,n){return this.delegate={iterator:j(t),resultName:e,nextLoc:n},"next"===this.method&&(this.arg=u),d}},i}(t.exports);try{regeneratorRuntime=e;}catch(t){Function("r","regeneratorRuntime = r")(e);}}),J=U(function(t){var r=Object.prototype.hasOwnProperty,h="~";function n(){}function s(t,e,n){this.fn=t,this.context=e,this.once=n||!1;}function o(t,e,n,r,o){if("function"!=typeof n)throw new TypeError("The listener must be a function");var i=new s(n,r||t,o),a=h?h+e:e;return t._events[a]?t._events[a].fn?t._events[a]=[t._events[a],i]:t._events[a].push(i):(t._events[a]=i,t._eventsCount++),t}function c(t,e){0==--t._eventsCount?t._events=new n:delete t._events[e];}function e(){this._events=new n,this._eventsCount=0;}Object.create&&(n.prototype=Object.create(null),(new n).__proto__||(h=!1)),e.prototype.eventNames=function(){var t,e,n=[];if(0===this._eventsCount)return n;for(e in t=this._events)r.call(t,e)&&n.push(h?e.slice(1):e);return Object.getOwnPropertySymbols?n.concat(Object.getOwnPropertySymbols(t)):n},e.prototype.listeners=function(t){var e=h?h+t:t,n=this._events[e];if(!n)return [];if(n.fn)return [n.fn];for(var r=0,o=n.length,i=new Array(o);r<o;r++)i[r]=n[r].fn;return i},e.prototype.listenerCount=function(t){var e=h?h+t:t,n=this._events[e];return n?n.fn?1:n.length:0},e.prototype.emit=function(t,e,n,r,o,i){var a=h?h+t:t;if(!this._events[a])return !1;var s,u,c=this._events[a],l=arguments.length;if(c.fn){switch(c.once&&this.removeListener(t,c.fn,void 0,!0),l){case 1:return c.fn.call(c.context),!0;case 2:return c.fn.call(c.context,e),!0;case 3:return c.fn.call(c.context,e,n),!0;case 4:return c.fn.call(c.context,e,n,r),!0;case 5:return c.fn.call(c.context,e,n,r,o),!0;case 6:return c.fn.call(c.context,e,n,r,o,i),!0}for(u=1,s=new Array(l-1);u<l;u++)s[u-1]=arguments[u];c.fn.apply(c.context,s);}else {var f,p=c.length;for(u=0;u<p;u++)switch(c[u].once&&this.removeListener(t,c[u].fn,void 0,!0),l){case 1:c[u].fn.call(c[u].context);break;case 2:c[u].fn.call(c[u].context,e);break;case 3:c[u].fn.call(c[u].context,e,n);break;case 4:c[u].fn.call(c[u].context,e,n,r);break;default:if(!s)for(f=1,s=new Array(l-1);f<l;f++)s[f-1]=arguments[f];c[u].fn.apply(c[u].context,s);}}return !0},e.prototype.on=function(t,e,n){return o(this,t,e,n,!1)},e.prototype.once=function(t,e,n){return o(this,t,e,n,!0)},e.prototype.removeListener=function(t,e,n,r){var o=h?h+t:t;if(!this._events[o])return this;if(!e)return c(this,o),this;var i=this._events[o];if(i.fn)i.fn!==e||r&&!i.once||n&&i.context!==n||c(this,o);else {for(var a=0,s=[],u=i.length;a<u;a++)(i[a].fn!==e||r&&!i[a].once||n&&i[a].context!==n)&&s.push(i[a]);s.length?this._events[o]=1===s.length?s[0]:s:c(this,o);}return this},e.prototype.removeAllListeners=function(t){var e;return t?(e=h?h+t:t,this._events[e]&&c(this,e)):(this._events=new n,this._eventsCount=0),this},e.prototype.off=e.prototype.removeListener,e.prototype.addListener=e.prototype.on,e.prefixed=h,e.EventEmitter=e,t.exports=e;}),W=function(){function n(t){var e;return o(this,n),(e=p(this,c(n).call(this,t))).name="TimeoutError",e}return u(n,e(Error)),n}(),Y=B,V=B,$=W;Y.default=V,Y.TimeoutError=$;var K=U(function(t,e){Object.defineProperty(e,"__esModule",{value:!0}),e.default=function(t,e,n){for(var r=0,o=t.length;0<o;){var i=o/2|0,a=r+i;n(t[a],e)<=0?(r=++a,o-=1+i):o=i;}return r};});Q(K);var X=U(function(t,e){Object.defineProperty(e,"__esModule",{value:!0});var n=function(){function t(){o(this,t),Object.defineProperty(this,"_queue",{enumerable:!0,configurable:!0,writable:!0,value:[]});}return i(t,[{key:"enqueue",value:function(t,e){var n={priority:(e=a({priority:0},e)).priority,run:t};if(this.size&&this._queue[this.size-1].priority>=e.priority)this._queue.push(n);else {var r=K.default(this._queue,n,function(t,e){return e.priority-t.priority});this._queue.splice(r,0,n);}}},{key:"dequeue",value:function(){var t=this._queue.shift();return t&&t.run}},{key:"size",get:function(){return this._queue.length}}]),t}();e.default=n;});Q(X);var Z=Q(U(function(t,e){Object.defineProperty(e,"__esModule",{value:!0});function r(){}var s=new Y.TimeoutError,n=function(){function n(t){var e;if(o(this,n),e=p(this,c(n).call(this)),Object.defineProperty(f(e),"_carryoverConcurrencyCount",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_isIntervalIgnored",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_intervalCount",{enumerable:!0,configurable:!0,writable:!0,value:0}),Object.defineProperty(f(e),"_intervalCap",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_interval",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_intervalEnd",{enumerable:!0,configurable:!0,writable:!0,value:0}),Object.defineProperty(f(e),"_intervalId",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_timeoutId",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_queue",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_queueClass",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_pendingCount",{enumerable:!0,configurable:!0,writable:!0,value:0}),Object.defineProperty(f(e),"_concurrency",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_isPaused",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_resolveEmpty",{enumerable:!0,configurable:!0,writable:!0,value:r}),Object.defineProperty(f(e),"_resolveIdle",{enumerable:!0,configurable:!0,writable:!0,value:r}),Object.defineProperty(f(e),"_timeout",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),Object.defineProperty(f(e),"_throwOnTimeout",{enumerable:!0,configurable:!0,writable:!0,value:void 0}),!("number"==typeof(t=a({carryoverConcurrencyCount:!1,intervalCap:1/0,interval:0,concurrency:1/0,autoStart:!0,queueClass:X.default},t)).intervalCap&&1<=t.intervalCap))throw new TypeError("Expected `intervalCap` to be a number from 1 and up, got `".concat(t.intervalCap,"` (").concat(I(t.intervalCap),")"));if(void 0===t.interval||!(Number.isFinite(t.interval)&&0<=t.interval))throw new TypeError("Expected `interval` to be a finite number >= 0, got `".concat(t.interval,"` (").concat(I(t.interval),")"));return e._carryoverConcurrencyCount=t.carryoverConcurrencyCount,e._isIntervalIgnored=t.intervalCap===1/0||0===t.interval,e._intervalCap=t.intervalCap,e._interval=t.interval,e._queue=new t.queueClass,e._queueClass=t.queueClass,e.concurrency=t.concurrency,e._timeout=t.timeout,e._throwOnTimeout=!0===t.throwOnTimeout,e._isPaused=!1===t.autoStart,e}return u(n,J),i(n,[{key:"_next",value:function(){this._pendingCount--,this._tryToStartAnother();}},{key:"_resolvePromises",value:function(){this._resolveEmpty(),this._resolveEmpty=r,0===this._pendingCount&&(this._resolveIdle(),this._resolveIdle=r);}},{key:"_onResumeInterval",value:function(){this._onInterval(),this._initializeIntervalIfNeeded(),this._timeoutId=void 0;}},{key:"_isIntervalPaused",value:function(){var t=this,e=Date.now();if(void 0===this._intervalId){var n=this._intervalEnd-e;if(!(n<0))return void 0===this._timeoutId&&(this._timeoutId=setTimeout(function(){t._onResumeInterval();},n)),!0;this._intervalCount=this._carryoverConcurrencyCount?this._pendingCount:0;}return !1}},{key:"_tryToStartAnother",value:function(){if(0===this._queue.size)return this._intervalId&&clearInterval(this._intervalId),this._intervalId=void 0,this._resolvePromises(),!1;if(!this._isPaused){var t=!this._isIntervalPaused();if(this._doesIntervalAllowAnother&&this._doesConcurrentAllowAnother)return this.emit("active"),this._queue.dequeue()(),t&&this._initializeIntervalIfNeeded(),!0}return !1}},{key:"_initializeIntervalIfNeeded",value:function(){var t=this;this._isIntervalIgnored||void 0!==this._intervalId||(this._intervalId=setInterval(function(){t._onInterval();},this._interval),this._intervalEnd=Date.now()+this._interval);}},{key:"_onInterval",value:function(){0===this._intervalCount&&0===this._pendingCount&&this._intervalId&&(clearInterval(this._intervalId),this._intervalId=void 0),this._intervalCount=this._carryoverConcurrencyCount?this._pendingCount:0,this._processQueue();}},{key:"_processQueue",value:function(){for(;this._tryToStartAnother(););}},{key:"add",value:function(o){var i,a=this,e=arguments;return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:return i=1<e.length&&void 0!==e[1]?e[1]:{},t.abrupt("return",new Promise(function(n,r){a._queue.enqueue(function(){var e;return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:return a._pendingCount++,a._intervalCount++,t.prev=2,e=void 0===a._timeout&&void 0===i.timeout?o():Y.default(Promise.resolve(o()),void 0===i.timeout?a._timeout:i.timeout,function(){(void 0===i.throwOnTimeout?a._throwOnTimeout:i.throwOnTimeout)&&r(s);}),t.t0=n,t.next=7,H.awrap(e);case 7:t.t1=t.sent,(0, t.t0)(t.t1),t.next=14;break;case 11:t.prev=11,t.t2=t.catch(2),r(t.t2);case 14:a._next();case 15:case"end":return t.stop()}},null,null,[[2,11]])},i),a._tryToStartAnother();}));case 2:case"end":return t.stop()}})}},{key:"addAll",value:function(e,n){var r=this;return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:return t.abrupt("return",Promise.all(e.map(function(e){return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:return t.abrupt("return",r.add(e,n));case 1:case"end":return t.stop()}})})));case 1:case"end":return t.stop()}})}},{key:"start",value:function(){return this._isPaused&&(this._isPaused=!1,this._processQueue()),this}},{key:"pause",value:function(){this._isPaused=!0;}},{key:"clear",value:function(){this._queue=new this._queueClass;}},{key:"onEmpty",value:function(){var n=this;return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:if(0===this._queue.size)return t.abrupt("return");t.next=2;break;case 2:return t.abrupt("return",new Promise(function(t){var e=n._resolveEmpty;n._resolveEmpty=function(){e(),t();};}));case 3:case"end":return t.stop()}},null,this)}},{key:"onIdle",value:function(){var n=this;return H.async(function(t){for(;;)switch(t.prev=t.next){case 0:if(0===this._pendingCount&&0===this._queue.size)return t.abrupt("return");t.next=2;break;case 2:return t.abrupt("return",new Promise(function(t){var e=n._resolveIdle;n._resolveIdle=function(){e(),t();};}));case 3:case"end":return t.stop()}},null,this)}},{key:"_doesIntervalAllowAnother",get:function(){return this._isIntervalIgnored||this._intervalCount<this._intervalCap}},{key:"_doesConcurrentAllowAnother",get:function(){return this._pendingCount<this._concurrency}},{key:"concurrency",get:function(){return this._concurrency},set:function(t){if(!("number"==typeof t&&1<=t))throw new TypeError("Expected `concurrency` to be a number from 1 and up, got `".concat(t,"` (").concat(I(t),")"));this._concurrency=t,this._processQueue();}},{key:"size",get:function(){return this._queue.size}},{key:"pending",get:function(){return this._pendingCount}},{key:"isPaused",get:function(){return this._isPaused}},{key:"timeout",set:function(t){this._timeout=t;},get:function(){return this._timeout}}]),n}();e.default=n;})),et=(nt.prototype.postEvent=function(t){var e=h(h({},t),this._getCommonParams());this.post(this._url,e);},nt.prototype.getEvent=function(t){var e=S(h(h({},t),this._getCommonParams())),n=this._url+"?"+e;this.get(n);},nt.prototype.post=function(r,o){var t=this;return new Promise(function(e,n){t.queue.add(function(){return _(t,void 0,void 0,function(){return d(this,function(t){switch(t.label){case 0:return [4,this.doRequest(r,"POST",o).then(e).catch(n)];case 1:return t.sent(),[2]}})})});})},nt.prototype.get=function(r){var t=this;return new Promise(function(e,n){t.queue.add(function(){return _(t,void 0,void 0,function(){return d(this,function(t){switch(t.label){case 0:return [4,this.doRequest(r,"GET").then(e).catch(n)];case 1:return t.sent(),[2]}})})});})},nt.prototype._getUrl=function(){return "https://"+(this._options.reportDomain||M)+"/log/sentry/v2/api/slardar/main/"},nt.prototype._getCommonParams=function(){return {timestamp:Date.now()}},nt.prototype.doRequest=function(r,o,i){return _(this,void 0,void 0,function(){return d(this,function(t){return [2,new Promise(function(e,n){R({url:r,method:o,header:{"Content-Type":"text/plain;charset=UTF-8"},data:i,success:function(t){e(t);},fail:function(t){n(t.errMsg);}});})]})})},nt);function nt(t){this.queue=new Z({concurrency:1}),this._options=t||{},this._url=this._getUrl();}function rt(){return y(Date)?Math.round(Date.now()/1e3):0}function ot(t,n){if(!m(t))return {};if(!y(n))return {};var r={};return P(t,function(t,e){n(e)&&(r[t]=e);}),r}function it(t){return g(t.value)?{metrics_type:"log",event_name:"default",log_content:t.value,category:ot(t.tags,g),timestamp:rt()}:{}}var at,st,ut,ct=(n(st=ft,ut=at=x),st.prototype=null===ut?Object.create(ut):(lt.prototype=ut.prototype,new lt),ft.prototype.setupTransport=function(){var t=new et({reportDomain:this._options.commonParams.report_domain});at.prototype.setupTransport.call(this,t);},ft.prototype._send=function(t){var e=t.event||{};e=h(h({},e),this._normalizeCommonParams(this._options.commonParams||{})),this._options.custom&&y(this._options.custom.beforeSend)&&!(e=this._options.custom.beforeSend(e))||("get"!==t.type?"post"===t.type&&this._transport.postEvent(e):this._transport.getEvent(e));},ft.prototype._sendPageview=function(){y(this._sendEvent)&&this._sendEvent({name:"PageView",type:"get",event:{ev_type:"pageview"}});},ft);function lt(){this.constructor=st;}function ft(t){var o=at.call(this,t)||this;return o.sendEvent=function(t){y(o._sendEvent)&&o._sendEvent(t);},o._shouldSend=function(t){return !!t.event},o._modifyEvent=function(t){return t},o._normalizeOptions=function(t){return O({},o._options,t)},o._normalizeCommonParams=function(t){var e=h({},t);return m(t.context)&&(e.context=JSON.stringify(t.context)),e||{}},o.updateConfig=function(t){var e=o._options.commonParams.pid;o._options=o._normalizeOptions(t),t.commonParams.pid!==e&&o._sendPageview();},o.sendPageview=function(){o._sendPageview();},o._emitSingleEvent=function(t){t.event_name&&y(o._sendEvent)&&o._sendEvent({name:"Emit",type:"post",event:{ev_type:"flexible",flexible_data_list:[t]}});},o.emitTimer=function(t){var e,n,r=g((e=t).name)&&w(e.value)?{metrics_type:"timer",event_name:"default",metrics:(n={},n[e.name]=e.value,n),category:ot(e.tags,g),timestamp:rt()}:{};o._emitSingleEvent(r);},o.emitCounter=function(t){var e,n,r=g((e=t).name)&&w(e.value)?{metrics_type:"counter",event_name:"default",metrics:(n={},n[e.name]=e.value,n),category:ot(e.tags,g),timestamp:rt()}:{};o._emitSingleEvent(r);},o.emitLog=function(t){var e=it(t);o._emitSingleEvent(e);},o.emitCustom=function(t){var e,n=g((e=t).event_name)?{metrics_type:"custom",event_name:e.event_name,metrics:ot(e.metrics,w),category:ot(e.tags,g),timestamp:rt()}:{};o._emitSingleEvent(n);},o}var pt=(ht.prototype._start=function(){var a=this;G&&G(function(t){var e,n=t,r=n.message;n.stack&&n.message||(e={message:r});var o=Date.now();if(a._errorMap.has(r)){var i=a._errorMap.get(r);if(i&&o-i<=5e3)return !1}a._errorMap.set(r,Date.now()),1e3<=a._errorMap.size&&a._errorMap.clear(),a._callback({name:a.name,type:"post",event:{ev_type:"jserr",timestamp:Date.now(),sentry:e,url:"https://127.0.0.1/index.html"}});});},ht);function ht(t){var e=this;this.name="TTErrorMonitor",this._enable=!1,this._errorMap=new Map,this.setup=function(t){e._enable?(e._callback=t,e._start()):console.log("[minigame_sdk] "+e.name+" monitor disabled.");},t&&t.enable&&(this._enable=t.enable);}function vt(n){if(!n)return "";var r=[];return Object.keys(n).forEach(function(t){var e=n[t];r.push(t+": "+e);}),r.join("\r\n")}function _t(t){var e=t.match(/^(https?:)\/\/(([^:/?#]*)(?::([0-9]+))?)([/]{0,1}[^?#]*)(\?[^#]*|)(#.*|)$/);return {protocol:"https",hostname:e[3],pathname:e[5]||"/"}}var dt=(mt.prototype._start=function(){this._overrideRequest();},mt.prototype._overrideRequest=function(){var i=this;tt.request=function(t){var e=t.success,n=t.fail,r=v(t,["success","fail"]),o=Date.now();return R(h({success:function(t){i._reportRequestSuccess(o,r,t),e&&e(t);},fail:function(t){i._reportRequestFail(o,r,t),n&&n(t);}},r))};},mt.prototype._getDataLength=function(t){return "string"==typeof t?t.length:"object"===I(t)?void 0!==t.byteLength?t.byteLength:Object.keys(t).length:0},mt.prototype._reportRequestSuccess=function(t,e,n){var r=Date.now(),o=e.url,i=e.method,a=void 0===i?"GET":i,s=e.header,u=n.statusCode,c=n.data,l=n.header,f=this._getDataLength(c),p=_t(o);p.hostname&&this._callback({name:this.name,type:"get",event:{ev_type:"ajax",ax_url:o,ax_status:u,ax_type:a,ax_request_header:vt(s),ax_duration:r-t,ax_size:f,ax_response_header:vt(l),ax_protocol:p.protocol,ax_domain:p.hostname,ax_path:p.pathname}});},mt.prototype._reportRequestFail=function(t,e,n){var r=Date.now(),o=e.url,i=e.method,a=void 0===i?"GET":i,s=n.errMsg,u=it({value:"[tt.request fail]\n"+a+": "+o+"\nduration:"+(r-t)+"\n"+s,tags:{monitor:"tt-request-fail"}});this._callback({name:this.name,type:"post",event:{ev_type:"flexible",flexible_data_list:[u]}});},mt);function mt(t){var e=this;this.name="TTRequestMonitor",this._enable=!1,this.setup=function(t){e._enable?(e._callback=t,e._start()):console.log("[minigame_sdk] "+e.name+" monitor disabled.");},t&&t.enable&&(this._enable=t.enable);}var yt=(gt.prototype._start=function(){this._overrideDownloadFile();},gt.prototype._overrideDownloadFile=function(){var i=this;tt.downloadFile=function(t){var e=t.success,n=t.fail,r=v(t,["success","fail"]),o=Date.now();return z(h({success:function(t){i._reportDownloadFileSuccess(o,r,t),e&&e(t);},fail:function(t){i._reportDownloadFileFail(o,r,t),n&&n(t);}},r))};},gt.prototype._reportDownloadFileSuccess=function(t,e,n){var r=Date.now(),o=e.url,i=e.header,a=n.statusCode,s=_t(o);s.hostname&&this._callback({name:this.name,type:"get",event:{ev_type:"ajax",ax_url:o,ax_status:a,ax_type:"GET",ax_request_header:vt(i),ax_duration:r-t,ax_size:1,ax_response_header:"",ax_protocol:s.protocol,ax_domain:s.hostname,ax_path:s.pathname}});},gt.prototype._reportDownloadFileFail=function(t,e,n){var r=Date.now(),o=e.url,i=n.errMsg,a=_t(o);if(a.hostname){this._callback({name:this.name,type:"get",event:{ev_type:"static",st_type:"file",st_src:o,st_protocol:a.protocol,st_domain:a.hostname,st_path:a.pathname}});var s=it({value:"[tt.downloadfile fail]\nGET: "+o+"\nduration:"+(r-t)+"\n"+i,tags:{monitor:"tt-downloadfile-fail"}});this._callback({name:this.name,type:"post",event:{ev_type:"flexible",flexible_data_list:[s]}});}},gt);function gt(t){var e=this;this.name="TTDownloadFileMonitor",this._enable=!1,this.setup=function(t){e._enable?(e._callback=t,e._start()):console.log("[minigame_sdk] "+e.name+" monitor disabled.");},t&&t.enable&&(this._enable=t.enable);}var bt=Object.prototype.hasOwnProperty,wt="~";function xt(){}function Ct(t,e,n){this.fn=t,this.context=e,this.once=n||!1;}function Pt(t,e,n,r,o){if("function"!=typeof n)throw new TypeError("The listener must be a function");var i=new Ct(n,r||t,o),a=wt?wt+e:e;return t._events[a]?t._events[a].fn?t._events[a]=[t._events[a],i]:t._events[a].push(i):(t._events[a]=i,t._eventsCount++),t}function Ot(t,e){0==--t._eventsCount?t._events=new xt:delete t._events[e];}function Et(){this._events=new xt,this._eventsCount=0;}Object.create&&(xt.prototype=Object.create(null),(new xt).__proto__||(wt=!1)),Et.prototype.eventNames=function(){var t,e,n=[];if(0===this._eventsCount)return n;for(e in t=this._events)bt.call(t,e)&&n.push(wt?e.slice(1):e);return Object.getOwnPropertySymbols?n.concat(Object.getOwnPropertySymbols(t)):n},Et.prototype.listeners=function(t){var e=wt?wt+t:t,n=this._events[e];if(!n)return [];if(n.fn)return [n.fn];for(var r=0,o=n.length,i=new Array(o);r<o;r++)i[r]=n[r].fn;return i},Et.prototype.listenerCount=function(t){var e=wt?wt+t:t,n=this._events[e];return n?n.fn?1:n.length:0},Et.prototype.emit=function(t,e,n,r,o,i){var a=wt?wt+t:t;if(!this._events[a])return !1;var s,u,c=this._events[a],l=arguments.length;if(c.fn){switch(c.once&&this.removeListener(t,c.fn,void 0,!0),l){case 1:return c.fn.call(c.context),!0;case 2:return c.fn.call(c.context,e),!0;case 3:return c.fn.call(c.context,e,n),!0;case 4:return c.fn.call(c.context,e,n,r),!0;case 5:return c.fn.call(c.context,e,n,r,o),!0;case 6:return c.fn.call(c.context,e,n,r,o,i),!0}for(u=1,s=new Array(l-1);u<l;u++)s[u-1]=arguments[u];c.fn.apply(c.context,s);}else {var f,p=c.length;for(u=0;u<p;u++)switch(c[u].once&&this.removeListener(t,c[u].fn,void 0,!0),l){case 1:c[u].fn.call(c[u].context);break;case 2:c[u].fn.call(c[u].context,e);break;case 3:c[u].fn.call(c[u].context,e,n);break;case 4:c[u].fn.call(c[u].context,e,n,r);break;default:if(!s)for(f=1,s=new Array(l-1);f<l;f++)s[f-1]=arguments[f];c[u].fn.apply(c[u].context,s);}}return !0},Et.prototype.on=function(t,e,n){return Pt(this,t,e,n,!1)},Et.prototype.once=function(t,e,n){return Pt(this,t,e,n,!0)},Et.prototype.removeListener=function(t,e,n,r){var o=wt?wt+t:t;if(!this._events[o])return this;if(!e)return Ot(this,o),this;var i=this._events[o];if(i.fn)i.fn!==e||r&&!i.once||n&&i.context!==n||Ot(this,o);else {for(var a=0,s=[],u=i.length;a<u;a++)(i[a].fn!==e||r&&!i[a].once||n&&i[a].context!==n)&&s.push(i[a]);s.length?this._events[o]=1===s.length?s[0]:s:Ot(this,o);}return this},Et.prototype.removeAllListeners=function(t){var e;return t?(e=wt?wt+t:t,this._events[e]&&Ot(this,e)):(this._events=new xt,this._eventsCount=0),this},Et.prototype.off=Et.prototype.removeListener,Et.prototype.addListener=Et.prototype.on,Et.prefixed=wt;var St=new(Et.EventEmitter=Et),jt=(kt.prototype._start=function(){this._overrideLogin();},kt.prototype._overrideLogin=function(){var i=this;tt.login=function(t){var e=t.success,n=t.fail,r=v(t,["success","fail"]),o=Date.now();return N(h({success:function(t){i._reportLoginSuccess(o,r,t),e&&e(t);},fail:function(t){i._reportLoginFail(o,r,t),n&&n(t);}},r))};},kt.prototype._reportLoginSuccess=function(t,e,n){var r=Date.now(),o=e.force,i=n.anonymousCode;this._callback({name:this.name,type:"get",event:{ev_type:"login",login_duration:r-t,login_force:o,login_anonymousCode:i}});},kt.prototype._reportLoginFail=function(t,e,n){var r=Date.now(),o=e.force,i=n.errMsg,a=it({value:"[tt.login fail]\nforce: "+o+"\nduration:"+(r-t)+"\n"+i,tags:{monitor:"tt-login-fail"}});this._callback({name:this.name,type:"post",event:{ev_type:"flexible",flexible_data_list:[a]}});},kt);function kt(t){var e=this;this.name="TTLoginMonitor",this._enable=!1,this.setup=function(t){e._enable?(e._callback=t,e._start()):console.log("[minigame_sdk] "+e.name+" monitor disabled.");},t&&t.enable&&(this._enable=t.enable);}var It=function(){var i=this;this.version="1.0.0",this._setBaseParams=function(){i.baseParams.bid=i.shared.config.commonParams.bid||"",i.baseParams.pid=i.shared.config.commonParams.pid||"";},this.instance=function(){for(var t=[],e=0;e<arguments.length;e++)t[e]=arguments[e];var n=t[0];if("config"!==n){if(i._client&&"emit"===n){if(!t[1]||!t[2])return;"timer"===t[1]&&i._client.emitTimer(t[2]),"counter"===t[1]&&i._client.emitCounter(t[2]),"log"===t[1]&&i._client.emitLog(t[2]),"custom"===t[1]&&i._client.emitCustom(t[2]);}}else {var r=t[1];i._configSetting=r;var o=i._normalizeSetting(r);i._client?i._client.updateConfig(o):i._init(o);}},this._init=function(t){i.shared.monitors={};var e=[new pt(t.monitors.tt_error_monitor),new dt(t.monitors.tt_request_monitor),new yt(t.monitors.tt_downloadfile_monitor),new jt(t.monitors.tt_login_monitor)],n=h(h({},t),{defaultMonitors:e});i._client=new ct(n),i.shared.instance&&b(i.shared.instance)&&i.shared.instance.push(n.commonParams.bid),i._initClient(i._baseClientConfig);},this._initClient=function(t){var e=i._normalizeSetting(t);i._client.updateConfig(e),Math.random()<e.commonParams.sample_rate&&(i._client.setupTransport(),i._client.sendPageview(),i.shared.transport=i._client._transport,i.shared.sendEvent=i._client.sendEvent);},this._normalizeSetting=function(t){void 0===t&&(t=i._baseClientConfig);var e=i._getClientCommonParams(t),n=i._getClientMonitorsSetting(t),r=i._getClientCustom(t),o=O(i._baseClientConfig,{commonParams:e,monitors:n,custom:r},{commonParams:t.commonParams||{},monitors:t.monitors||{},custom:t.custom||{}},{commonParams:i._getClientCommonParams(i._configSetting),monitors:i._getClientMonitorsSetting(i._configSetting),custom:i._getClientCustom(i._configSetting)});return i.shared.config=o,i._setBaseParams(),i._baseClientConfig=o},this._getClientCommonParams=function(t){var n={},r=["bid","context","pid","sample_rate","slardar_session_id","slardar_web_id"];return P(t,function(t,e){!function(t,e){if(!b(t))return 0;if(0===t.length)return 0;for(var n=0;n<t.length;){if(t[n]===e)return 1;n++;}return 0}(r,t)||(n[t]=e);}),t.cookieid&&(n.slardar_web_id=t.cookieid),t.sampleRate&&(n.sample_rate=t.sampleRate),t.reportDomain&&(n.report_domain=t.reportDomain),t.domain&&(n.report_domain=t.domain),n.slardar_web_id&&(n.slardar_web_id=n.slardar_web_id.toString()),n},this._getClientCustom=function(t){var e={};return y(t.beforeSend)&&(e.beforeSend=t.beforeSend),e},this._getClientMonitorsSetting=function(t){return h(h({},i._baseClientConfig.monitors),t.monitors||{})},this._baseClientConfig={defaultMonitors:[],commonParams:{version:this.version,hostname:"127.0.0.1",protocol:"https",url:"http://127.0.0.1/index.html",slardar_session_id:D,sample_rate:1,pid:"index.html",report_domain:M,screen_resolution:F()&&window.tt.getSystemInfoSync().screenHeight&&window.tt.getSystemInfoSync().screenWidth?window.tt.getSystemInfoSync().screenWidth+"x"+window.tt.getSystemInfoSync().screenHeight:"0x0",network_type:"unknown",bid:"",context:{},slardar_web_id:A},monitors:{tt_error_monitor:{enable:!0},tt_request_monitor:{enable:!0},tt_downloadfile_monitor:{enable:!0},tt_login_monitor:{enable:!0}},custom:{beforeSend:function(t){return t}}},this._client=void 0,this._configSetting={},this.shared={config:this._baseClientConfig,transport:this._client?this._client._transport:void 0,emitter:St,instance:[],plugins:{},monitors:{},sendEvent:void 0},this.baseParams={bid:this.shared.config.commonParams.bid||"",pid:this.shared.config.commonParams.pid||""};};var Tt=function(t){};if(F()){var Lt=new It;(Tt=Lt.instance).version=Lt.version,Tt.shared=Lt.shared,Tt.SlardarMiniGame=It,Tt._baseParams=Lt.baseParams;}else console.log("[slardar-minigame]: Current runtime not support tt mini-program sdk.");var qt=Tt;qt.shared&&(qt.shared.installedType="npm"),t.SlardarMiniGame=It,t.default=qt,Object.defineProperty(t,"__esModule",{value:!0});});
});

var slardar = unwrapExports(lib);

/* eslint-disable @typescript-eslint/camelcase */
var AIInfos = [
    {
        nickName: "海纳百川嘿嘿",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/29c5054ec55541078423d522fc35d0cf",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "困困了",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小曹在吃饭",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱你&",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "A在路上18634",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/91be63b83cffdea848c3330ade42b533",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "笑面",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "斜杠青年是我",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "玫瑰花海",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "点点点点",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "农村来的孩子",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户23451",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "承诺3391",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "隔壁老汪",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/9aaf893a2dece2dff55b82a616c70385",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "燕子菲菲",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户271275",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "骑猪上高速888",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/2e8439eae11f804001da48609300b1d4",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "风轻云淡",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "大大大金钢",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小菜源",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "精灵之水",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/519b8421a6c3a6049614dc4e8fff1dce",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱你到永远",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户4390231",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "欢欢和乐乐",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "火山小公民",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/833424c376a478d496c3d54de1433ade",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "种桃山",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "刘自在",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.jpg",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "三农学校",
        avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "长发飘飘",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/e961a64fd0355960ba3e9922ed153278",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "君子兰和一枝梅",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.png",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "无心法师",
        avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.webp",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "心诚则灵哦",
        avatarUrl: "https://weili.pstatp.com/weili/sm/426848925334503725.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "冰要融化了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/623228522403070134.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "转角错过了你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/553477068105580572.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心一百天",
        avatarUrl: "https://weili.pstatp.com/weili/sm/240089462197190945.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户934214",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587648334714896394.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户6680453",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669616943813362018.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "长春烤冷面大师",
        avatarUrl: "https://weili.pstatp.com/weili/sm/658420230293946378.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "叫姐姐",
        avatarUrl: "https://weili.pstatp.com/weili/sm/428516205869006917.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "山野",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587446892231000357.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "南城旧梦8123",
        avatarUrl: "https://weili.pstatp.com/weili/sm/384307231256215560.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "我爱洗澡",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587229592655626693.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心过好每一天779",
        avatarUrl: "https://weili.pstatp.com/weili/sm/467280991544541423.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阜阳141",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587302976466846012.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "幸福满满140",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669632689163468848.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户30702357",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587405188098556325.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "梁凉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587297066591846784.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心61405039",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587205051212497417.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阿布都-热合曼",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669634106502676910.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "孩子的天下彤彤",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587267422727569815.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "你的丹宝吖?",
        avatarUrl: "https://weili.pstatp.com/weili/sm/462528103423148101.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "?娅熹妹?",
        avatarUrl: "https://weili.pstatp.com/weili/sm/454181633347289219.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "阳光28548346",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669554520758681820.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "微笑天天341",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587253842040979803.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户22463029",
        avatarUrl: "https://weili.pstatp.com/weili/sm/250591619290759212.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一切随风413",
        avatarUrl: "https://weili.pstatp.com/weili/sm/449418703594586223.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "狼行天下6666……",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587200051870564464.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "きお征服一切",
        avatarUrl: "https://weili.pstatp.com/weili/sm/342637355472650273.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "再见只是陌路人",
        avatarUrl: "https://weili.pstatp.com/weili/sm/446882688554893315.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木头人123",
        avatarUrl: "https://weili.pstatp.com/weili/sm/545342451586564123.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "雨墨墨哥哥",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587716556047646881.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "眼镜20190304",
        avatarUrl: "https://weili.pstatp.com/weili/sm/455005837571391582.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "小哥哥小哥哥",
        avatarUrl: "https://weili.pstatp.com/weili/sm/256835256216453164.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "记忆中的你2838",
        avatarUrl: "https://weili.pstatp.com/weili/sm/535044185163956250.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "虞美人的晚霞",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587677952881590544.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户65011007",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587442322385797480.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱钓鱼和吃虾",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587200215079321785.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "鹏鹏",
        avatarUrl: "https://weili.pstatp.com/weili/sm/662748749878132739.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一只小飞侠",
        avatarUrl: "https://weili.pstatp.com/weili/sm/587258240087490915.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "珍爱一生",
        avatarUrl: "https://weili.pstatp.com/weili/sm/601065365875196168.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "何必要这样",
        avatarUrl: "https://weili.pstatp.com/weili/sm/669651629969244198.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "我是你的小迷妹",
        avatarUrl: "https://weili.pstatp.com/weili/sm/467259224650285091.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "老男孩12345",
        avatarUrl: "https://weili.pstatp.com/weili/sm/652397371889549317.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "想家的天天",
        avatarUrl: "https://weili.pstatp.com/weili/sm/646430476903514114.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "北漂23456",
        avatarUrl: "https://weili.pstatp.com/weili/sm/503537499081670669.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "影子91",
        avatarUrl: "https://weili.pstatp.com/weili/sm/529663570392121354.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "春梦随我飞",
        avatarUrl: "https://weili.pstatp.com/weili/sm/237053212376694855.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "AAA",
        avatarUrl: "https://weili.pstatp.com/weili/sm/431201917579689985.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "DAR",
        avatarUrl: "https://weili.pstatp.com/weili/sm/538631470695186531.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户43670985",
        avatarUrl: "https://weili.pstatp.com/weili/sm/583932036957733026.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "用户16888150",
        avatarUrl: "https://weili.pstatp.com/weili/sm/247417140260634798.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "福282",
        avatarUrl: "https://weili.pstatp.com/weili/sm/563093439754928337.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "发哥天天发",
        avatarUrl: "https://weili.pstatp.com/weili/sm/519550932457881629.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "又来了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/447455783281230332.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "谁懂姐的心",
        avatarUrl: "https://weili.pstatp.com/weili/sm/469145531281440785.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木子81060",
        avatarUrl: "https://weili.pstatp.com/weili/sm/624045725830480155.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "user946329",
        avatarUrl: "https://weili.pstatp.com/weili/sm/554989850074284301.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "彼岸微凉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/535046461497671779.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "A偶然与巧合",
        avatarUrl: "https://weili.pstatp.com/weili/sm/555735224466931864.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "无情波斯喵",
        avatarUrl: "https://weili.pstatp.com/weili/sm/415721266307530761.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "木美儿",
        avatarUrl: "https://weili.pstatp.com/weili/sm/508763202967699481.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "开心开心开心极了",
        avatarUrl: "https://weili.pstatp.com/weili/sm/495360233532686349.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "★触不可及",
        avatarUrl: "https://weili.pstatp.com/weili/sm/268082624513441827.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "苗文玉",
        avatarUrl: "https://weili.pstatp.com/weili/sm/589110118248349828.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "爱爱你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/628507294907891883.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "萍水相逢1234",
        avatarUrl: "https://weili.pstatp.com/weili/sm/664556407126097929.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "改变生活",
        avatarUrl: "https://weili.pstatp.com/weili/sm/580935171982229708.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "Www33",
        avatarUrl: "https://weili.pstatp.com/weili/sm/646969907616415745.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "青涩的雨",
        avatarUrl: "https://weili.pstatp.com/weili/sm/381388989957275648.jpg?app_id=wi9fe0c5a58ce0fecd",
        gender: 0,
        city: "",
        province: "",
        country: "",
        language: "",
        open_id: "",
        combatScore: 0,
        playTime: 0,
    },
    {
        nickName: "一笑而过是你",
        avatarUrl: "https://weili.pstatp.com/weili/sm/658167626086875143.jpg?app_id=wi9fe0c5a58ce0fecd"
    }
];

var config = {
    gameId: '',
    hostAppId: '',
    environment: 'test',
    isMultiPlayer: false,
    userInfo: {
        nickname: '',
        avatar: '',
        gender: 0,
        isAnchor: true,
        openId: '',
    },
};
var startMatchTimestamp;
var matchTask;
var matchId = 'match-1';
var _debugInstance = null;
var setDebugLog = function (debugInstance) {
    _debugInstance = debugInstance;
};
var log = function () {
    var param = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        param[_i] = arguments[_i];
    }
    if (_debugInstance) {
        _debugInstance.log(param);
    }
};
// 重试机会
var retryChance = 0;
/**
 * 原因
 */
var REASON_CODE;
(function (REASON_CODE) {
    REASON_CODE[REASON_CODE["MATCH_SUCCESS"] = 0] = "MATCH_SUCCESS";
    REASON_CODE[REASON_CODE["MATCH_ERROR"] = 1] = "MATCH_ERROR";
    REASON_CODE[REASON_CODE["MATCH_TIMEOUT"] = 2] = "MATCH_TIMEOUT";
    REASON_CODE[REASON_CODE["MATCH_CANCEL"] = 3] = "MATCH_CANCEL";
    REASON_CODE[REASON_CODE["MATCH_CLOSE"] = 4] = "MATCH_CLOSE";
})(REASON_CODE || (REASON_CODE = {}));
/**
 * 初始化
 * @param gameId 游戏appid
 * @param hostAppId 宿主id
 * @param environment 环境
 * @param isMultiPlayer 是否多人游戏（非1v1），默认为1v1，仅用于埋点统计！
 * @param userInfo 玩家信息
 */
function init(param) {
    var gameId = param.gameId, hostAppId = param.hostAppId, environment = param.environment, userInfo = param.userInfo, isMultiPlayer = param.isMultiPlayer;
    config.gameId = gameId;
    config.hostAppId = hostAppId;
    config.environment = environment;
    config.isMultiPlayer = isMultiPlayer;
    config.userInfo = userInfo;
    pvpClient.PVPClient.initConfig({
        appId: config.gameId,
        aId: config.hostAppId,
        environment: environment === 'online' ? 'online' : 'test',
        platformId: 'wonderland',
        extraInfo: __assign(__assign({}, userInfo), { openid: userInfo.openId, getInnerMatchInfo: true, businessID: '2', isAnchor: userInfo.isAnchor, name: userInfo.nickName, icon: userInfo.avatarUrl }),
    });
    // 配置最基本的信息
    // NOTE: 引入 slardar 后，完成配置
    // DOC: https://slardar.bytedance.net/docs/115/325/28930/
    slardar('config', {
        bid: 'pvp_match_for_w_live',
        context: {
            name: 'w-pvp-match',
            env: config.environment,
        },
        // eslint-disable-next-line @typescript-eslint/camelcase
        slardar_web_id: config.userInfo.openID,
    });
}
/**
 * 开始匹配
 * @param retryCount 重试次数
 */
function startMatch(retryCount // 重试次数;
) {
    var _this = this;
    var lock = false;
    retryChance = retryCount;
    switch (config.environment) {
        case 'develop':
            matchId = 'match-1';
            break;
        case 'test':
            matchId = 'match-2';
            break;
        case 'online':
            matchId = 'match-3';
            break;
    }
    matchTask = pvpClient.PVPClient.startMatch({
        matchId: matchId,
        disableAutoRetry: true,
    });
    var defer = new Deferred();
    matchTask.once('cancel', function (param) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            if (!lock) {
                log('cancel match');
                // 匹配超时一次，重新匹配
                if (param && param.code === errorCode.ERROR_CODE.MATCH_TIME_OUT) {
                    log('retryChance timeout', retryChance);
                    restartMatch(defer, { reason: REASON_CODE.MATCH_TIMEOUT });
                }
                else {
                    defer.reject({ reason: REASON_CODE.MATCH_CANCEL });
                    addMatchEndSlardar('cancel');
                }
                lock = true;
            }
            return [2 /*return*/];
        });
    }); });
    matchTask.once('start', function () { return __awaiter(_this, void 0, void 0, function () {
        var reportData;
        return __generator(this, function (_a) {
            log('start match');
            // 匹配开始，打点，计时
            slardar('emit', 'counter', {
                // 统计次数相关的打点
                name: 'mp_match_start',
                value: 1,
                tags: {
                    mp_id: config.gameId,
                    _param_for_special: 'micro_game',
                    launch_from: 'live',
                    location: 'com_component',
                    scene: '023012',
                },
            });
            reportData = {
                mp_id: config.gameId,
                _param_for_special: 'micro_game',
                launch_from: 'live',
                location: 'com_component',
                scene: '023012',
            };
            // @ts-ignore
            if (typeof tt !== 'undefined') {
                // @ts-ignore
                tt.reportAnalytics('mp_match_start', reportData, 1);
            }
            startMatchTimestamp = new Date().getTime();
            return [2 /*return*/];
        });
    }); });
    matchTask.once('error', function (param) {
        if (!lock) {
            log('error match', param);
            restartMatch(defer, {
                reason: REASON_CODE.MATCH_ERROR,
                status: param,
            });
            lock = true;
        }
    });
    matchTask.once('success', function (param) {
        lock = true;
        var enemyType = 'player';
        if (config.isMultiPlayer)
            enemyType = 'no_sense';
        else {
            if (param.playerList.length < 2) {
                enemyType = 'ai';
            }
            param.playerList.forEach(function (player) {
                if (!player.isMe) {
                    if (player.extraInfo.isAnchor) {
                        enemyType = 'host';
                    }
                    else {
                        enemyType = 'player';
                    }
                }
            });
        }
        addMatchEndSlardar('success', enemyType);
        // 修改playerList.extraInfo中openid为openId
        var playerList = [];
        (param.playerList || []).forEach(function (player) {
            var newPlayer = __assign({}, player);
            Object.keys(player.extraInfo).forEach(function (key) {
                if (key === 'openid') {
                    newPlayer.extraInfo.openId = player.extraInfo[key];
                }
                else if (key === 'name') {
                    // 兼容来自大厅的name字段
                    newPlayer.extraInfo.nickName = player.extraInfo[key];
                }
                else if (key === 'icon') {
                    // 兼容来自大厅的icon字段
                    newPlayer.extraInfo.avatarUrl = player.extraInfo[key];
                }
                else {
                    newPlayer.extraInfo[key] = player.extraInfo[key];
                }
            });
            playerList.push(newPlayer);
        });
        defer.resolve({ roomNum: param.roomNum, playerList: playerList });
    });
    matchTask.once('close', function (param) {
        if (!lock) {
            log('close match');
            restartMatch(defer, {
                reason: REASON_CODE.MATCH_CLOSE,
                status: param,
            });
            lock = true;
        }
    });
    return defer.promise;
}
/**
 * 取消匹配
 */
function cancelMatch() {
    if (matchTask) {
        matchTask.cancel();
    }
}
// 获取指定数量的随机元素
function getRandomArrayElements(arr, count) {
    var ret = [];
    for (var i = 0; i < count; i++) {
        var index = Math.floor(Math.random() * arr.length);
        ret.push(arr.splice(index, 1)[0]);
    }
    return ret;
}
/**
 * 获取指定数量的ai信息
 * @param count 数量
 */
function getAIInfo(count) {
    if (count <= 0) {
        return [];
    }
    var AIInfoList = getRandomArrayElements(AIInfos.slice(), count);
    return AIInfoList;
}
/**
 * 第一次匹配之后的重试，用于第一次超时或匹配失败后的自动重新匹配
 */
function restartMatch(defer, lastRejectParam) {
    var lock = false;
    log('retryChance', retryChance);
    if (retryChance === 0) {
        var resultType = void 0;
        switch (lastRejectParam.reason) {
            case REASON_CODE.MATCH_TIMEOUT:
                resultType = 'expired';
                break;
            case REASON_CODE.MATCH_ERROR:
            case REASON_CODE.MATCH_CLOSE:
                resultType = 'fail';
                break;
            case REASON_CODE.MATCH_CANCEL:
                resultType = 'cancel';
                break;
            default:
                resultType = 'fail';
                break;
        }
        addMatchEndSlardar(resultType);
        return defer.reject(lastRejectParam);
    }
    retryChance--;
    matchTask = pvpClient.PVPClient.startMatch({
        matchId: matchId,
        disableAutoRetry: true,
    });
    matchTask.once('cancel', function (param) {
        if (!lock) {
            lock = true;
            log('cancel restart');
            var code = void 0;
            if (param && param.code === errorCode.ERROR_CODE.MATCH_TIME_OUT) {
                code = REASON_CODE.MATCH_TIMEOUT;
                addMatchEndSlardar('expired');
            }
            else {
                code = REASON_CODE.MATCH_CANCEL;
                addMatchEndSlardar('cancel');
            }
            defer.reject({ reason: code });
        }
    });
    matchTask.once('error', function (param) {
        if (!lock) {
            log('error restart', param);
            restartMatch(defer, {
                reason: REASON_CODE.MATCH_ERROR,
                status: param,
            });
            lock = true;
        }
    });
    matchTask.once('close', function (param) {
        if (!lock) {
            log('close restart');
            restartMatch(defer, {
                reason: REASON_CODE.MATCH_CLOSE,
                status: param,
            });
            lock = true;
        }
    });
    matchTask.once('success', function (param) {
        lock = true;
        var enemyType = 'player';
        if (config.isMultiPlayer)
            enemyType = 'no_sense';
        else {
            if (param.playerList.length < 2) {
                enemyType = 'ai';
            }
            param.playerList.forEach(function (player) {
                if (!player.isMe) {
                    if (player.extraInfo.isAnchor) {
                        enemyType = 'host';
                    }
                    else {
                        enemyType = 'player';
                    }
                }
            });
        }
        addMatchEndSlardar('success', enemyType);
        // 修改playerList中openid为openId
        var playerList = [];
        (param.playerList || []).forEach(function (player) {
            var newPlayer = __assign({}, player);
            Object.keys(player.extraInfo).forEach(function (key) {
                if (key === 'openid') {
                    newPlayer.extraInfo.openId = player.extraInfo[key];
                }
                else if (key === 'name') {
                    // 兼容来自大厅的name字段
                    newPlayer.extraInfo.nickName = player.extraInfo[key];
                }
                else if (key === 'icon') {
                    // 兼容来自大厅的icon字段
                    newPlayer.extraInfo.avatarUrl = player.extraInfo[key];
                }
                else {
                    newPlayer.extraInfo[key] = player.extraInfo[key];
                }
            });
            playerList.push(newPlayer);
        });
        defer.resolve({ roomNum: param.roomNum, playerList: playerList });
    });
    return defer.promise;
}
function addMatchEndSlardar(resultType, enemyType) {
    if (enemyType === void 0) { enemyType = ''; }
    slardar('emit', 'timer', {
        // 统计时间相关的打点
        name: 'mp_match_result',
        value: new Date().getTime() - startMatchTimestamp,
        tags: {
            mp_id: config.gameId,
            _param_for_special: 'micro_game',
            launch_from: 'live',
            location: 'com_component',
            scene: '023012',
            enemy_type: enemyType,
            result_type: resultType,
        },
    });
    var reportData = {
        mp_id: config.gameId,
        _param_for_special: 'micro_game',
        launch_from: 'live',
        location: 'com_component',
        scene: '023012',
        duration: new Date().getTime() - startMatchTimestamp,
        enemy_type: enemyType,
        result_type: resultType,
    };
    // @ts-ignore
    if (typeof tt !== 'undefined') {
        // @ts-ignore
        tt.reportAnalytics('mp_match_result', reportData, 1);
    }
    startMatchTimestamp = 0;
}

var api = /*#__PURE__*/Object.freeze({
    __proto__: null,
    setDebugLog: setDebugLog,
    log: log,
    get REASON_CODE () { return REASON_CODE; },
    init: init,
    startMatch: startMatch,
    cancelMatch: cancelMatch,
    getAIInfo: getAIInfo
});

exports.PVPMatch = api;

}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
