// @byted-creative/pvp_match/dist/pvp_match.cjs.development.js
(function() {"use strict";var __module = CC_EDITOR ? module : {exports:{}};var __filename = 'preview-scripts/__node_modules/@byted-creative/pvp_match/dist/pvp_match.cjs.development.js';var __require = CC_EDITOR ? function (request) {return cc.require(request, require);} : function (request) {return cc.require(request, __filename);};function __define (exports, require, module) {cc._RF.push(module, '', 'pvp_match.cjs.development', __filename);'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var pvpClient = require('@byted-creative/pvp-client');
var errorCode = require('@byted-creative/pvp-client/build/common/errorCode');
var slardar = _interopDefault(require('@byted-creative/app_log'));

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

/**
 * 延迟承诺类。属于Promise的衍生类。
 */
var Deferred =
/*#__PURE__*/

/** @class */
function () {
  function Deferred() {
    var _this = this;

    this.promise = new Promise(function (resolve, reject) {
      Object.defineProperties(_this, {
        resolve: {
          value: function (value) {
            // @ts-ignore
            _this.value = value;
            resolve(value);
          },
          writable: false
        },
        reject: {
          value: reject,
          writable: false
        }
      });
    });
  }

  return Deferred;
}();

/* eslint-disable @typescript-eslint/camelcase */
var AIInfos = [{
  nickName: "海纳百川嘿嘿",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/29c5054ec55541078423d522fc35d0cf",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "困困了",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "小曹在吃饭",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "爱你&",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/29c5054ec55541078423d522fc35d0cf~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "A在路上18634",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/91be63b83cffdea848c3330ade42b533",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "笑面",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "斜杠青年是我",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "玫瑰花海",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/91be63b83cffdea848c3330ade42b533~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "点点点点",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "农村来的孩子",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户23451",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "承诺3391",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/f3bfa41a8e7804bbd3d551a3dd068df4~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "隔壁老汪",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/9aaf893a2dece2dff55b82a616c70385",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "燕子菲菲",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户271275",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/9aaf893a2dece2dff55b82a616c70385~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "骑猪上高速888",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/2e8439eae11f804001da48609300b1d4",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "风轻云淡",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "大大大金钢",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "小菜源",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/2e8439eae11f804001da48609300b1d4~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "精灵之水",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/519b8421a6c3a6049614dc4e8fff1dce",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "爱你到永远",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户4390231",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "欢欢和乐乐",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/519b8421a6c3a6049614dc4e8fff1dce~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "火山小公民",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/obj/motor-img/833424c376a478d496c3d54de1433ade",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "种桃山",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "刘自在",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.jpg",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "三农学校",
  avatarUrl: "https://sf3-ttcdn-tos.pstatp.com/img/motor-img/833424c376a478d496c3d54de1433ade~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "长发飘飘",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/obj/motor-img/e961a64fd0355960ba3e9922ed153278",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "君子兰和一枝梅",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.png",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "无心法师",
  avatarUrl: "https://sf1-ttcdn-tos.pstatp.com/img/motor-img/e961a64fd0355960ba3e9922ed153278~noop.webp",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "心诚则灵哦",
  avatarUrl: "https://weili.pstatp.com/weili/sm/426848925334503725.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "冰要融化了",
  avatarUrl: "https://weili.pstatp.com/weili/sm/623228522403070134.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "转角错过了你",
  avatarUrl: "https://weili.pstatp.com/weili/sm/553477068105580572.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "开心一百天",
  avatarUrl: "https://weili.pstatp.com/weili/sm/240089462197190945.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户934214",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587648334714896394.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户6680453",
  avatarUrl: "https://weili.pstatp.com/weili/sm/669616943813362018.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "长春烤冷面大师",
  avatarUrl: "https://weili.pstatp.com/weili/sm/658420230293946378.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "叫姐姐",
  avatarUrl: "https://weili.pstatp.com/weili/sm/428516205869006917.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "山野",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587446892231000357.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "南城旧梦8123",
  avatarUrl: "https://weili.pstatp.com/weili/sm/384307231256215560.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "我爱洗澡",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587229592655626693.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "开心过好每一天779",
  avatarUrl: "https://weili.pstatp.com/weili/sm/467280991544541423.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "阜阳141",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587302976466846012.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "幸福满满140",
  avatarUrl: "https://weili.pstatp.com/weili/sm/669632689163468848.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户30702357",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587405188098556325.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "梁凉",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587297066591846784.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "开心61405039",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587205051212497417.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "阿布都-热合曼",
  avatarUrl: "https://weili.pstatp.com/weili/sm/669634106502676910.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "孩子的天下彤彤",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587267422727569815.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "你的丹宝吖?",
  avatarUrl: "https://weili.pstatp.com/weili/sm/462528103423148101.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "?娅熹妹?",
  avatarUrl: "https://weili.pstatp.com/weili/sm/454181633347289219.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "阳光28548346",
  avatarUrl: "https://weili.pstatp.com/weili/sm/669554520758681820.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "微笑天天341",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587253842040979803.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户22463029",
  avatarUrl: "https://weili.pstatp.com/weili/sm/250591619290759212.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "一切随风413",
  avatarUrl: "https://weili.pstatp.com/weili/sm/449418703594586223.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "狼行天下6666……",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587200051870564464.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "きお征服一切",
  avatarUrl: "https://weili.pstatp.com/weili/sm/342637355472650273.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "再见只是陌路人",
  avatarUrl: "https://weili.pstatp.com/weili/sm/446882688554893315.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "木头人123",
  avatarUrl: "https://weili.pstatp.com/weili/sm/545342451586564123.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "雨墨墨哥哥",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587716556047646881.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "眼镜20190304",
  avatarUrl: "https://weili.pstatp.com/weili/sm/455005837571391582.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "小哥哥小哥哥",
  avatarUrl: "https://weili.pstatp.com/weili/sm/256835256216453164.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "记忆中的你2838",
  avatarUrl: "https://weili.pstatp.com/weili/sm/535044185163956250.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "虞美人的晚霞",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587677952881590544.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户65011007",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587442322385797480.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "爱钓鱼和吃虾",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587200215079321785.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "鹏鹏",
  avatarUrl: "https://weili.pstatp.com/weili/sm/662748749878132739.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "一只小飞侠",
  avatarUrl: "https://weili.pstatp.com/weili/sm/587258240087490915.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "珍爱一生",
  avatarUrl: "https://weili.pstatp.com/weili/sm/601065365875196168.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "何必要这样",
  avatarUrl: "https://weili.pstatp.com/weili/sm/669651629969244198.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "我是你的小迷妹",
  avatarUrl: "https://weili.pstatp.com/weili/sm/467259224650285091.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "老男孩12345",
  avatarUrl: "https://weili.pstatp.com/weili/sm/652397371889549317.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "想家的天天",
  avatarUrl: "https://weili.pstatp.com/weili/sm/646430476903514114.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "北漂23456",
  avatarUrl: "https://weili.pstatp.com/weili/sm/503537499081670669.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "影子91",
  avatarUrl: "https://weili.pstatp.com/weili/sm/529663570392121354.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "春梦随我飞",
  avatarUrl: "https://weili.pstatp.com/weili/sm/237053212376694855.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "AAA",
  avatarUrl: "https://weili.pstatp.com/weili/sm/431201917579689985.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "DAR",
  avatarUrl: "https://weili.pstatp.com/weili/sm/538631470695186531.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户43670985",
  avatarUrl: "https://weili.pstatp.com/weili/sm/583932036957733026.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "用户16888150",
  avatarUrl: "https://weili.pstatp.com/weili/sm/247417140260634798.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "福282",
  avatarUrl: "https://weili.pstatp.com/weili/sm/563093439754928337.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "发哥天天发",
  avatarUrl: "https://weili.pstatp.com/weili/sm/519550932457881629.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "又来了",
  avatarUrl: "https://weili.pstatp.com/weili/sm/447455783281230332.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "谁懂姐的心",
  avatarUrl: "https://weili.pstatp.com/weili/sm/469145531281440785.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "木子81060",
  avatarUrl: "https://weili.pstatp.com/weili/sm/624045725830480155.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "user946329",
  avatarUrl: "https://weili.pstatp.com/weili/sm/554989850074284301.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "彼岸微凉",
  avatarUrl: "https://weili.pstatp.com/weili/sm/535046461497671779.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "A偶然与巧合",
  avatarUrl: "https://weili.pstatp.com/weili/sm/555735224466931864.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "无情波斯喵",
  avatarUrl: "https://weili.pstatp.com/weili/sm/415721266307530761.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "木美儿",
  avatarUrl: "https://weili.pstatp.com/weili/sm/508763202967699481.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "开心开心开心极了",
  avatarUrl: "https://weili.pstatp.com/weili/sm/495360233532686349.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "★触不可及",
  avatarUrl: "https://weili.pstatp.com/weili/sm/268082624513441827.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "苗文玉",
  avatarUrl: "https://weili.pstatp.com/weili/sm/589110118248349828.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "爱爱你",
  avatarUrl: "https://weili.pstatp.com/weili/sm/628507294907891883.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "萍水相逢1234",
  avatarUrl: "https://weili.pstatp.com/weili/sm/664556407126097929.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "改变生活",
  avatarUrl: "https://weili.pstatp.com/weili/sm/580935171982229708.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "Www33",
  avatarUrl: "https://weili.pstatp.com/weili/sm/646969907616415745.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "青涩的雨",
  avatarUrl: "https://weili.pstatp.com/weili/sm/381388989957275648.jpg?app_id=wi9fe0c5a58ce0fecd",
  gender: 0,
  city: "",
  province: "",
  country: "",
  language: "",
  open_id: "",
  combatScore: 0,
  playTime: 0
}, {
  nickName: "一笑而过是你",
  avatarUrl: "https://weili.pstatp.com/weili/sm/658167626086875143.jpg?app_id=wi9fe0c5a58ce0fecd"
}];

var config = {
  gameId: '',
  hostAppId: '',
  environment: 'test',
  isMultiPlayer: false,
  userInfo: {
    nickname: '',
    avatar: '',
    gender: 0,
    isAnchor: true,
    openId: ''
  }
};
var startMatchTimestamp;
var matchTask;
var matchId = 'match-1';
var _debugInstance = null;
var setDebugLog = function (debugInstance) {
  _debugInstance = debugInstance;
};
var log = function () {
  var param = [];

  for (var _i = 0; _i < arguments.length; _i++) {
    param[_i] = arguments[_i];
  }

  if (_debugInstance) {
    _debugInstance.log(param);
  }
}; // 重试机会

var retryChance = 0;
/**
 * 原因
 */

var REASON_CODE;

(function (REASON_CODE) {
  REASON_CODE[REASON_CODE["MATCH_SUCCESS"] = 0] = "MATCH_SUCCESS";
  REASON_CODE[REASON_CODE["MATCH_ERROR"] = 1] = "MATCH_ERROR";
  REASON_CODE[REASON_CODE["MATCH_TIMEOUT"] = 2] = "MATCH_TIMEOUT";
  REASON_CODE[REASON_CODE["MATCH_CANCEL"] = 3] = "MATCH_CANCEL";
  REASON_CODE[REASON_CODE["MATCH_CLOSE"] = 4] = "MATCH_CLOSE";
})(REASON_CODE || (REASON_CODE = {}));
/**
 * 初始化
 * @param gameId 游戏appid
 * @param hostAppId 宿主id
 * @param environment 环境
 * @param isMultiPlayer 是否多人游戏（非1v1），默认为1v1，仅用于埋点统计！
 * @param userInfo 玩家信息
 */


function init(param) {
  var gameId = param.gameId,
      hostAppId = param.hostAppId,
      environment = param.environment,
      userInfo = param.userInfo,
      isMultiPlayer = param.isMultiPlayer;
  config.gameId = gameId;
  config.hostAppId = hostAppId;
  config.environment = environment;
  config.isMultiPlayer = isMultiPlayer;
  config.userInfo = userInfo;
  pvpClient.PVPClient.initConfig({
    appId: config.gameId,
    aId: config.hostAppId,
    environment: environment === 'online' ? 'online' : 'test',
    platformId: 'wonderland',
    extraInfo: __assign(__assign({}, userInfo), {
      openid: userInfo.openId,
      getInnerMatchInfo: true,
      businessID: '2',
      isAnchor: userInfo.isAnchor,
      name: userInfo.nickName,
      icon: userInfo.avatarUrl
    })
  }); // 配置最基本的信息
  // NOTE: 引入 slardar 后，完成配置
  // DOC: https://slardar.bytedance.net/docs/115/325/28930/

  slardar('config', {
    bid: 'pvp_match_for_w_live',
    context: {
      name: 'w-pvp-match',
      env: config.environment
    },
    // eslint-disable-next-line @typescript-eslint/camelcase
    slardar_web_id: config.userInfo.openID
  });
}
/**
 * 开始匹配
 * @param retryCount 重试次数
 */

function startMatch(retryCount // 重试次数;
) {
  var _this = this;

  var lock = false;
  retryChance = retryCount;

  switch (config.environment) {
    case 'develop':
      matchId = 'match-1';
      break;

    case 'test':
      matchId = 'match-2';
      break;

    case 'online':
      matchId = 'match-3';
      break;
  }

  matchTask = pvpClient.PVPClient.startMatch({
    matchId: matchId,
    disableAutoRetry: true
  });
  var defer = new Deferred();
  matchTask.once('cancel', function (param) {
    return __awaiter(_this, void 0, void 0, function () {
      return __generator(this, function (_a) {
        if (!lock) {
          log('cancel match'); // 匹配超时一次，重新匹配

          if (param && param.code === errorCode.ERROR_CODE.MATCH_TIME_OUT) {
            log('retryChance timeout', retryChance);
            restartMatch(defer, {
              reason: REASON_CODE.MATCH_TIMEOUT
            });
          } else {
            defer.reject({
              reason: REASON_CODE.MATCH_CANCEL
            });
            addMatchEndSlardar('cancel');
          }

          lock = true;
        }

        return [2
        /*return*/
        ];
      });
    });
  });
  matchTask.once('start', function () {
    return __awaiter(_this, void 0, void 0, function () {
      var reportData;
      return __generator(this, function (_a) {
        log('start match'); // 匹配开始，打点，计时

        slardar('emit', 'counter', {
          // 统计次数相关的打点
          name: 'mp_match_start',
          value: 1,
          tags: {
            mp_id: config.gameId,
            _param_for_special: 'micro_game',
            launch_from: 'live',
            location: 'com_component',
            scene: '023012'
          }
        });
        reportData = {
          mp_id: config.gameId,
          _param_for_special: 'micro_game',
          launch_from: 'live',
          location: 'com_component',
          scene: '023012'
        }; // @ts-ignore

        tt.reportAnalytics('mp_match_start', reportData, 1);
        startMatchTimestamp = new Date().getTime();
        return [2
        /*return*/
        ];
      });
    });
  });
  matchTask.once('error', function (param) {
    if (!lock) {
      log('error match', param);
      restartMatch(defer, {
        reason: REASON_CODE.MATCH_ERROR,
        status: param
      });
      lock = true;
    }
  });
  matchTask.once('success', function (param) {
    lock = true;
    var enemyType = 'player';
    if (config.isMultiPlayer) enemyType = 'no_sense';else {
      if (param.playerList.length < 2) {
        enemyType = 'ai';
      }

      param.playerList.forEach(function (player) {
        if (!player.isMe) {
          if (player.extraInfo.isAnchor) {
            enemyType = 'host';
          } else {
            enemyType = 'player';
          }
        }
      });
    }
    addMatchEndSlardar('success', enemyType); // 修改playerList.extraInfo中openid为openId

    var playerList = [];
    (param.playerList || []).forEach(function (player) {
      var newPlayer = __assign({}, player);

      Object.keys(player.extraInfo).forEach(function (key) {
        if (key === 'openid') {
          newPlayer.extraInfo.openId = player.extraInfo[key];
        } else if (key === 'name') {
          // 兼容来自大厅的name字段
          newPlayer.extraInfo.nickName = player.extraInfo[key];
        } else if (key === 'icon') {
          // 兼容来自大厅的icon字段
          newPlayer.extraInfo.avatarUrl = player.extraInfo[key];
        } else {
          newPlayer.extraInfo[key] = player.extraInfo[key];
        }
      });
      playerList.push(newPlayer);
    });
    defer.resolve({
      roomNum: param.roomNum,
      playerList: playerList
    });
  });
  matchTask.once('close', function (param) {
    if (!lock) {
      log('close match');
      restartMatch(defer, {
        reason: REASON_CODE.MATCH_CLOSE,
        status: param
      });
      lock = true;
    }
  });
  return defer.promise;
}
/**
 * 取消匹配
 */

function cancelMatch() {
  if (matchTask) {
    matchTask.cancel();
  }
} // 获取指定数量的随机元素

function getRandomArrayElements(arr, count) {
  var ret = [];

  for (var i = 0; i < count; i++) {
    var index = Math.floor(Math.random() * arr.length);
    ret.push(arr.splice(index, 1)[0]);
  }

  return ret;
}
/**
 * 获取指定数量的ai信息
 * @param count 数量
 */


function getAIInfo(count) {
  if (count <= 0) {
    return [];
  }

  var AIInfoList = getRandomArrayElements(AIInfos.slice(), count);
  return AIInfoList;
}
/**
 * 第一次匹配之后的重试，用于第一次超时或匹配失败后的自动重新匹配
 */

function restartMatch(defer, lastRejectParam) {
  var lock = false;
  log('retryChance', retryChance);

  if (retryChance === 0) {
    var resultType = void 0;

    switch (lastRejectParam.reason) {
      case REASON_CODE.MATCH_TIMEOUT:
        resultType = 'expired';
        break;

      case REASON_CODE.MATCH_ERROR:
      case REASON_CODE.MATCH_CLOSE:
        resultType = 'fail';
        break;

      case REASON_CODE.MATCH_CANCEL:
        resultType = 'cancel';
        break;

      default:
        resultType = 'fail';
        break;
    }

    addMatchEndSlardar(resultType);
    return defer.reject(lastRejectParam);
  }

  retryChance--;
  matchTask = pvpClient.PVPClient.startMatch({
    matchId: matchId,
    disableAutoRetry: true
  });
  matchTask.once('cancel', function (param) {
    if (!lock) {
      lock = true;
      log('cancel restart');
      var code = void 0;

      if (param && param.code === errorCode.ERROR_CODE.MATCH_TIME_OUT) {
        code = REASON_CODE.MATCH_TIMEOUT;
        addMatchEndSlardar('expired');
      } else {
        code = REASON_CODE.MATCH_CANCEL;
        addMatchEndSlardar('cancel');
      }

      defer.reject({
        reason: code
      });
    }
  });
  matchTask.once('error', function (param) {
    if (!lock) {
      log('error restart', param);
      restartMatch(defer, {
        reason: REASON_CODE.MATCH_ERROR,
        status: param
      });
      lock = true;
    }
  });
  matchTask.once('close', function (param) {
    if (!lock) {
      log('close restart');
      restartMatch(defer, {
        reason: REASON_CODE.MATCH_CLOSE,
        status: param
      });
      lock = true;
    }
  });
  matchTask.once('success', function (param) {
    lock = true;
    var enemyType = 'player';
    if (config.isMultiPlayer) enemyType = 'no_sense';else {
      if (param.playerList.length < 2) {
        enemyType = 'ai';
      }

      param.playerList.forEach(function (player) {
        if (!player.isMe) {
          if (player.extraInfo.isAnchor) {
            enemyType = 'host';
          } else {
            enemyType = 'player';
          }
        }
      });
    }
    addMatchEndSlardar('success', enemyType); // 修改playerList中openid为openId

    var playerList = [];
    (param.playerList || []).forEach(function (player) {
      var newPlayer = __assign({}, player);

      Object.keys(player.extraInfo).forEach(function (key) {
        if (key === 'openid') {
          newPlayer.extraInfo.openId = player.extraInfo[key];
        } else if (key === 'name') {
          // 兼容来自大厅的name字段
          newPlayer.extraInfo.nickName = player.extraInfo[key];
        } else if (key === 'icon') {
          // 兼容来自大厅的icon字段
          newPlayer.extraInfo.avatarUrl = player.extraInfo[key];
        } else {
          newPlayer.extraInfo[key] = player.extraInfo[key];
        }
      });
      playerList.push(newPlayer);
    });
    defer.resolve({
      roomNum: param.roomNum,
      playerList: playerList
    });
  });
  return defer.promise;
}

function addMatchEndSlardar(resultType, enemyType) {
  if (enemyType === void 0) {
    enemyType = '';
  }

  slardar('emit', 'timer', {
    // 统计时间相关的打点
    name: 'mp_match_result',
    value: new Date().getTime() - startMatchTimestamp,
    tags: {
      mp_id: config.gameId,
      _param_for_special: 'micro_game',
      launch_from: 'live',
      location: 'com_component',
      scene: '023012',
      enemy_type: enemyType,
      result_type: resultType
    }
  });
  var reportData = {
    mp_id: config.gameId,
    _param_for_special: 'micro_game',
    launch_from: 'live',
    location: 'com_component',
    scene: '023012',
    duration: new Date().getTime() - startMatchTimestamp,
    enemy_type: enemyType,
    result_type: resultType
  }; // @ts-ignore

  tt.reportAnalytics('mp_match_result', reportData, 1);
  startMatchTimestamp = 0;
}

var api = {
    __proto__: null,
    setDebugLog: setDebugLog,
    log: log,
    get REASON_CODE () { return REASON_CODE; },
    init: init,
    startMatch: startMatch,
    cancelMatch: cancelMatch,
    getAIInfo: getAIInfo
};

exports.PVPMatch = api;

cc._RF.pop();
        }
        if (CC_EDITOR) {
            __define(__module.exports, __require, __module);
        }
        else {
            cc.registerModuleFunc(__filename, function () {
                __define(__module.exports, __require, __module);
            });
        }
        })();
